{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "feb5b243",
   "metadata": {},
   "source": [
    "## AnnData Tutorial\n",
    "\n",
    "AnnData is a python package that provides a datastructure for matrix like data that needs indexed annotations as well as unstructured metadata. It is used with ScanPy for analysing scRNA-seq data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a93fc28c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.12.2\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from scipy.sparse import csr_matrix\n",
    "print(ad.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86e8874e",
   "metadata": {},
   "source": [
    "### Creating and reading data\n",
    "Here we create some arbitrary random count data using the `csr_matrix` module. This is just a standin for some sparse count data like for example gene expression in a sequencing data set where we have many cells/nuclei that each have a small number of counts of many genes.\n",
    "\n",
    "We then read the data into a structured AnnData object, where each row becomes an observation and each column becomes a variable. Like cells with many expressed genes (gene is the variable)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "893633b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 100 × 2000"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating some data with the csr_matrix module\n",
    "counts = csr_matrix(np.random.poisson(1, size= (100,2000)), dtype= np.float32)\n",
    "\n",
    "# Reading it into an AnnData structure\n",
    "adata = ad.AnnData(counts)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b40fda72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
       "\twith 126451 stored elements and shape (100, 2000)>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# X is the part of the AnnData structure that contains the actual tabular locked raw data, the counts in this case\n",
    "adata.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57e31eff",
   "metadata": {},
   "source": [
    "## Indices\n",
    "\n",
    "Each dimension has indices in the AnnData object. Those, in our case, are the cell IDs and the Gene IDs. So we can reference each single cell and we can, for each single gene, read how many of the genes were expressed in this cell.\n",
    "\n",
    "The observations (cells) can be accessed via the `.obs` attribute and its associated functions and the variables can be accessed and changed under the `.var` attribute.\n",
    "\n",
    "Here we used the `.obs_names` and the `.var_names` methods to generate a name for each observation and variable in our artifical data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4cb53e36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Cell_0', 'Cell_1', 'Cell_2', 'Cell_3', 'Cell_4', 'Cell_5', 'Cell_6',\n",
      "       'Cell_7', 'Cell_8', 'Cell_9'],\n",
      "      dtype='object')\n",
      "Index(['Gene_0', 'Gene_1', 'Gene_2', 'Gene_3', 'Gene_4', 'Gene_5', 'Gene_6',\n",
      "       'Gene_7', 'Gene_8', 'Gene_9'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Making some cell and gene names\n",
    "adata.obs_names = [f\"Cell_{i:d}\" for i in range(adata.n_obs)] # We use the number of observations (rows of adata) for the list length\n",
    "adata.var_names = [f\"Gene_{i:d}\" for i in range(adata.n_vars)] # We use the number of variables (columns of adata)for the list length\n",
    "\n",
    "print(adata.obs_names[:10])\n",
    "print(adata.var_names[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2434485",
   "metadata": {},
   "source": [
    "### Subsetting AnnData\n",
    "\n",
    "We can now use the names of observations and variables as indices to subset and slice the AnnData object however we want!\n",
    "\n",
    "Cool for later analysis to, for example, pull out all Genes with a certain name (if you have a list of gene names belonging to a certain family) or to reference a set of specially annotated cells (foreshadowing). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3dd08f80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 2 × 2"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Suppose we are interest of Genes 5,1501, and 7 of cells 5 and 60\n",
    "adata[[\"Cell_5\", \"Cell_60\"], [\"Gene_5\", \"Gene_1501\", \"Gene_7\"]] # First rows, then columns\n",
    "# If you reference by key names, they have to go into lists\n",
    "\n",
    "# Or cells 4 to 59 and genes 300 to 600\n",
    "adata[4:60, 300:600]\n",
    "\n",
    "adata[[\"Cell_5\", \"Cell_6\"], [\"Gene_99\", \"Gene_100\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8817c22d",
   "metadata": {},
   "source": [
    "### Adding aligned metadata\n",
    "\n",
    "We between the \"column\" that has the row names (cell IDs) and the row that has the column names (gene IDs here) there can be an limitless number of other descriptive __Annotation Layers__ that contain metadata like alternative gene names, cell barcodes, plate origin, batch ID, known gene origin (like mito genes for example).\n",
    "\n",
    "Those annotation layers are regular Pandas DataFrames. Dis y we need Pandus, gud ting I did the Pandus Tootrial lately heheee.\n",
    "\n",
    "Here we use `np.random` again to generate a list of random categorical data containing fictitious cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e9c5dc4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cell_0</th>\n",
       "      <td>IHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_1</th>\n",
       "      <td>OHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_2</th>\n",
       "      <td>Pillar Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_3</th>\n",
       "      <td>IHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_4</th>\n",
       "      <td>OHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_95</th>\n",
       "      <td>Pillar Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_96</th>\n",
       "      <td>OHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_97</th>\n",
       "      <td>Fibrocyte_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_98</th>\n",
       "      <td>OHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_99</th>\n",
       "      <td>OHC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_type\n",
       "Cell_0           IHC\n",
       "Cell_1           OHC\n",
       "Cell_2   Pillar Cell\n",
       "Cell_3           IHC\n",
       "Cell_4           OHC\n",
       "...              ...\n",
       "Cell_95  Pillar Cell\n",
       "Cell_96          OHC\n",
       "Cell_97  Fibrocyte_3\n",
       "Cell_98          OHC\n",
       "Cell_99          OHC\n",
       "\n",
       "[100 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generating a list with random cell types the length of AnnData vertical dimension (observations)\n",
    "cell_types = np.random.choice([\"IHC\", \"OHC\", \"Pillar Cell\", \"Fibrocyte_3\"], size= (adata.n_obs,))\n",
    "\n",
    "adata.obs[\"cell_type\"] = pd.Categorical(cell_types) # We use categoricals pandas dataframe because more efficient\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2416f25",
   "metadata": {},
   "source": [
    "Eeeeh wolla! We got an extra column that tells us the cell type of each cell, sweet, neat, pristeet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e8a4afc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 100 × 2000\n",
       "    obs: 'cell_type'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata # Mind that the observations have changed! Theres now a new column"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71463a59",
   "metadata": {},
   "source": [
    "### Subsetting Metadata\n",
    "\n",
    "Now you can also subset all the observations by any annotation layer. Like filtering out all the cells belonging to a certain cell type.\n",
    "\n",
    "Because an AnnData object, like a Pandas df, or a regular ass Python list, also accepts boolean lists for indexing. You have to rereference the original data tho, hence why you have to put the adata also in the index.\n",
    "\n",
    "We can slap in a logical expression and filter through an annotation layer. If theres only one argument btw in the brackets, its always referring to rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aedb7aaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 52 × 2000\n",
       "    obs: 'cell_type'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[adata.obs.cell_type.str.contains(\"HC\")] # Filter out all the Hair Cells for example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "667ff4ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 25 × 1\n",
       "    obs: 'cell_type'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # OR\n",
    "adata[adata.obs.cell_type == \"Pillar Cell\", \"Gene_9\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d21288",
   "metadata": {},
   "source": [
    "### Multidimensional Annotations\n",
    "\n",
    "Metadata can have more than one dimension too! Like if you have a UMAP projection. This is handled in the `.obsm` and `.varm` attributes. Using those, you can annotate whole matrices and identify them with keys.\n",
    "\n",
    "<font color='red'> **HOWEVER!!!** </font>\n",
    "\n",
    "The y-dimension of .obsm needs to match n_obs and the x-dimension of .varm needs to match n_vars."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad94a5d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AxisArrays with keys: X_umap"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We create a matrix\n",
    "n_dim_obs_annotation = np.random.normal(0, 1, size= (adata.n_obs, 2))\n",
    "n_dim_var_annotation = np.random.normal(0, 1, size= (adata.n_vars, 2))\n",
    "\n",
    "adata.obsm[\"X_umap\"] = n_dim_obs_annotation\n",
    "adata.varm[\"gene_things\"] = n_dim_var_annotation\n",
    "\n",
    "adata.varm\n",
    "adata.obsm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "68fcca7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 100 × 2000\n",
       "    obs: 'cell_type'\n",
       "    obsm: 'X_umap'\n",
       "    varm: 'gene_things'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94d6bfef",
   "metadata": {},
   "source": [
    "### Unstructured Metadata\n",
    "\n",
    "Theres a sparate little field in an AnnData object that neither corresponds to vars nor obs. It is for unstructured metadata. This can be the name of the sample or batch, it could be some experiment ID or the date or just a description of the analysis or whatever. It can have any structure, like a list or array or dataframe.\n",
    "\n",
    "IT CAN NOT BE A DICTIONARY! AT LEAST NOT IF YOU WANNA SAVE THE FILE TO .h5ad because it throws an error and I dont know why.\n",
    "\n",
    "Use the `.uns` attribute to access it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "58734a60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('randothing',\n",
       "              ['This is the experiment A', '2025-03-11', 'Johann was here'])])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating an unstructured metadata entry and filling it widdabuncha crepe\n",
    "adata.uns[\"randothing\"] = [\"This is the experiment A\", \"2025-03-11\", \"Johann was here\"]\n",
    "\n",
    "adata.uns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fb414cb",
   "metadata": {},
   "source": [
    "### Layers\n",
    "\n",
    "This is different from when I say \"annotation layer\". Annotations are just maximum 2D labels that go on the edges of our data matrix to tell us what is in those rows and columns and what their names are.\n",
    "A Layer here is a complete parallel copy of the whole AnnData matrix that is being slapped on top.\n",
    "\n",
    "This is useful to track analysis steps. If you transform the data for example you can save it to a new layer to preserve the pre-transformed stuff.\n",
    "\n",
    "**Note:** Not everything is stored in layers! Only the core data matrix X is contained in the new layer. So if you wanna do anything with the Layers, you always have to call `.X` after the layer indexing like this: `adata.layer[\"Layer_A\"].X`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0de440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 100 × 2000\n",
       "    obs: 'cell_type'\n",
       "    uns: 'randothing'\n",
       "    obsm: 'X_umap'\n",
       "    varm: 'gene_things'\n",
       "    layers: 'log_transform'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers[\"log_transform\"] = np.log1p(adata.X) # IMPORTANT! A layer only copies the matrix itself\n",
    "adata # See how theres a layer now!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5df675e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_0</th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_1990</th>\n",
       "      <th>Gene_1991</th>\n",
       "      <th>Gene_1992</th>\n",
       "      <th>Gene_1993</th>\n",
       "      <th>Gene_1994</th>\n",
       "      <th>Gene_1995</th>\n",
       "      <th>Gene_1996</th>\n",
       "      <th>Gene_1997</th>\n",
       "      <th>Gene_1998</th>\n",
       "      <th>Gene_1999</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cell_0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_95</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_96</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_97</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_98</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_99</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 2000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene_0  Gene_1  Gene_2  Gene_3  Gene_4  Gene_5  Gene_6  Gene_7  \\\n",
       "Cell_0      0.0     0.0     1.0     0.0     1.0     2.0     2.0     0.0   \n",
       "Cell_1      2.0     0.0     0.0     0.0     0.0     1.0     1.0     0.0   \n",
       "Cell_2      1.0     1.0     2.0     2.0     1.0     1.0     3.0     1.0   \n",
       "Cell_3      0.0     1.0     1.0     2.0     0.0     1.0     1.0     0.0   \n",
       "Cell_4      0.0     1.0     0.0     0.0     1.0     1.0     2.0     1.0   \n",
       "...         ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "Cell_95     1.0     0.0     2.0     1.0     1.0     1.0     0.0     1.0   \n",
       "Cell_96     1.0     3.0     0.0     0.0     3.0     0.0     3.0     1.0   \n",
       "Cell_97     0.0     0.0     1.0     2.0     1.0     2.0     3.0     1.0   \n",
       "Cell_98     2.0     2.0     0.0     0.0     1.0     2.0     1.0     0.0   \n",
       "Cell_99     1.0     2.0     0.0     0.0     0.0     0.0     1.0     0.0   \n",
       "\n",
       "         Gene_8  Gene_9  ...  Gene_1990  Gene_1991  Gene_1992  Gene_1993  \\\n",
       "Cell_0      0.0     0.0  ...        1.0        1.0        2.0        1.0   \n",
       "Cell_1      0.0     0.0  ...        0.0        0.0        1.0        1.0   \n",
       "Cell_2      0.0     1.0  ...        2.0        0.0        1.0        1.0   \n",
       "Cell_3      4.0     1.0  ...        1.0        1.0        2.0        0.0   \n",
       "Cell_4      0.0     1.0  ...        1.0        1.0        2.0        1.0   \n",
       "...         ...     ...  ...        ...        ...        ...        ...   \n",
       "Cell_95     4.0     2.0  ...        2.0        2.0        2.0        1.0   \n",
       "Cell_96     1.0     1.0  ...        0.0        2.0        2.0        2.0   \n",
       "Cell_97     0.0     1.0  ...        0.0        0.0        0.0        0.0   \n",
       "Cell_98     0.0     1.0  ...        1.0        1.0        0.0        1.0   \n",
       "Cell_99     0.0     0.0  ...        1.0        1.0        1.0        2.0   \n",
       "\n",
       "         Gene_1994  Gene_1995  Gene_1996  Gene_1997  Gene_1998  Gene_1999  \n",
       "Cell_0         0.0        1.0        1.0        1.0        0.0        3.0  \n",
       "Cell_1         1.0        2.0        2.0        2.0        0.0        0.0  \n",
       "Cell_2         0.0        0.0        0.0        0.0        1.0        0.0  \n",
       "Cell_3         3.0        0.0        0.0        2.0        1.0        1.0  \n",
       "Cell_4         4.0        0.0        1.0        2.0        0.0        0.0  \n",
       "...            ...        ...        ...        ...        ...        ...  \n",
       "Cell_95        1.0        1.0        4.0        0.0        0.0        0.0  \n",
       "Cell_96        0.0        1.0        0.0        0.0        0.0        1.0  \n",
       "Cell_97        1.0        3.0        2.0        2.0        1.0        0.0  \n",
       "Cell_98        1.0        2.0        1.0        0.0        1.0        2.0  \n",
       "Cell_99        1.0        0.0        1.0        2.0        0.0        0.0  \n",
       "\n",
       "[100 rows x 2000 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.to_df() # Look, regular ass normal sparse counts form the poisson method from numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "149b29f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_0</th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_1990</th>\n",
       "      <th>Gene_1991</th>\n",
       "      <th>Gene_1992</th>\n",
       "      <th>Gene_1993</th>\n",
       "      <th>Gene_1994</th>\n",
       "      <th>Gene_1995</th>\n",
       "      <th>Gene_1996</th>\n",
       "      <th>Gene_1997</th>\n",
       "      <th>Gene_1998</th>\n",
       "      <th>Gene_1999</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cell_0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.386294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_1</th>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_2</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>...</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.609438</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.609438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_95</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.609438</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>...</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.609438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_96</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_97</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_98</th>\n",
       "      <td>1.098612</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_99</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 2000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene_0    Gene_1    Gene_2    Gene_3    Gene_4    Gene_5    Gene_6  \\\n",
       "Cell_0   0.000000  0.000000  0.693147  0.000000  0.693147  1.098612  1.098612   \n",
       "Cell_1   1.098612  0.000000  0.000000  0.000000  0.000000  0.693147  0.693147   \n",
       "Cell_2   0.693147  0.693147  1.098612  1.098612  0.693147  0.693147  1.386294   \n",
       "Cell_3   0.000000  0.693147  0.693147  1.098612  0.000000  0.693147  0.693147   \n",
       "Cell_4   0.000000  0.693147  0.000000  0.000000  0.693147  0.693147  1.098612   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "Cell_95  0.693147  0.000000  1.098612  0.693147  0.693147  0.693147  0.000000   \n",
       "Cell_96  0.693147  1.386294  0.000000  0.000000  1.386294  0.000000  1.386294   \n",
       "Cell_97  0.000000  0.000000  0.693147  1.098612  0.693147  1.098612  1.386294   \n",
       "Cell_98  1.098612  1.098612  0.000000  0.000000  0.693147  1.098612  0.693147   \n",
       "Cell_99  0.693147  1.098612  0.000000  0.000000  0.000000  0.000000  0.693147   \n",
       "\n",
       "           Gene_7    Gene_8    Gene_9  ...  Gene_1990  Gene_1991  Gene_1992  \\\n",
       "Cell_0   0.000000  0.000000  0.000000  ...   0.693147   0.693147   1.098612   \n",
       "Cell_1   0.000000  0.000000  0.000000  ...   0.000000   0.000000   0.693147   \n",
       "Cell_2   0.693147  0.000000  0.693147  ...   1.098612   0.000000   0.693147   \n",
       "Cell_3   0.000000  1.609438  0.693147  ...   0.693147   0.693147   1.098612   \n",
       "Cell_4   0.693147  0.000000  0.693147  ...   0.693147   0.693147   1.098612   \n",
       "...           ...       ...       ...  ...        ...        ...        ...   \n",
       "Cell_95  0.693147  1.609438  1.098612  ...   1.098612   1.098612   1.098612   \n",
       "Cell_96  0.693147  0.693147  0.693147  ...   0.000000   1.098612   1.098612   \n",
       "Cell_97  0.693147  0.000000  0.693147  ...   0.000000   0.000000   0.000000   \n",
       "Cell_98  0.000000  0.000000  0.693147  ...   0.693147   0.693147   0.000000   \n",
       "Cell_99  0.000000  0.000000  0.000000  ...   0.693147   0.693147   0.693147   \n",
       "\n",
       "         Gene_1993  Gene_1994  Gene_1995  Gene_1996  Gene_1997  Gene_1998  \\\n",
       "Cell_0    0.693147   0.000000   0.693147   0.693147   0.693147   0.000000   \n",
       "Cell_1    0.693147   0.693147   1.098612   1.098612   1.098612   0.000000   \n",
       "Cell_2    0.693147   0.000000   0.000000   0.000000   0.000000   0.693147   \n",
       "Cell_3    0.000000   1.386294   0.000000   0.000000   1.098612   0.693147   \n",
       "Cell_4    0.693147   1.609438   0.000000   0.693147   1.098612   0.000000   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "Cell_95   0.693147   0.693147   0.693147   1.609438   0.000000   0.000000   \n",
       "Cell_96   1.098612   0.000000   0.693147   0.000000   0.000000   0.000000   \n",
       "Cell_97   0.000000   0.693147   1.386294   1.098612   1.098612   0.693147   \n",
       "Cell_98   0.693147   0.693147   1.098612   0.693147   0.000000   0.693147   \n",
       "Cell_99   1.098612   0.693147   0.000000   0.693147   1.098612   0.000000   \n",
       "\n",
       "         Gene_1999  \n",
       "Cell_0    1.386294  \n",
       "Cell_1    0.000000  \n",
       "Cell_2    0.000000  \n",
       "Cell_3    0.693147  \n",
       "Cell_4    0.000000  \n",
       "...            ...  \n",
       "Cell_95   0.000000  \n",
       "Cell_96   0.693147  \n",
       "Cell_97   0.000000  \n",
       "Cell_98   1.098612  \n",
       "Cell_99   0.000000  \n",
       "\n",
       "[100 rows x 2000 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.to_df(layer= \"log_transform\") # BEHOLD\n",
    "# IT HAS THE SAME VAR AND OBS NAMES N EVERYTHING!! BUT DATA IS DIFFERENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bf42b099",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
       "\twith 126451 stored elements and shape (100, 2000)>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers[\"log_transform\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28440cd",
   "metadata": {},
   "source": [
    "### Saving to disk\n",
    "\n",
    "Things like clustering take a long time and annotating is a bitch to repeat! Save your work!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0ef71cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns.clear()\n",
    "\n",
    "adata.write_h5ad(\"15months_Analysis3.h5ad\", compression= \"gzip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d539286",
   "metadata": {},
   "source": [
    "### Copying and Analysing Large Files\n",
    "\n",
    "All modifications on AnnData objects are in-place, so no copy is created and everything is added or overwritten.\n",
    "Furthermore, every time you access a AnnData object via indexing or slicing you actually only get whats called **a View** of adata. It does not store any actual data.\n",
    "\n",
    "If you want to retain a subset of adata for any purpose, you need to add `.copy()` to the end of the instantiation of the subset (assuming you save it to a new variable).\n",
    "\n",
    "There is also something called **\"auto-copy\"** though. So if you subset adata AND make a change to it at the same time, then the View will turn into an actual object that holds changed data and it does the copying on its own.\n",
    "\n",
    "Also! It is possible to access a file in part by reading it from storage by using the `backed = 'r'` parameter in the `ad.read('file.h5ad, backed= 'r')` method call. The file will then be open for reading just partially which saves memory.\n",
    "If you do this, you have to manually close it at the end tho using `adata.file.close()` OR just open it using `with`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5941639b",
   "metadata": {},
   "source": [
    "### Misc\n",
    "\n",
    "Remember, AnnData is Pandas based. You can use all the Pandas functions on the obs and vars because they are just Pandas dataframes. You can also use Pandas logical function like `isin()` when slicing.\n",
    "\n",
    "Okay have fun!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cf06d3f",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
