{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a6eee7f7",
   "metadata": {},
   "source": [
    "## Preprocessing and Analysis of snRNAseq Data\n",
    "\n",
    "In this notebook I do quality control, preprocessing, and data analysis of single nuclei sequencing data (gene count matrices) generated by the Smart-seq3xpress protocol. For this I am mainly using the ScanPy module.<br>\n",
    "**Data used here is of the Spiral Ganglion Aging Atlas**\n",
    "\n",
    "The workflow presented here contains:\n",
    "- Reading the data\n",
    "- Removing annotation layers and only reatining some selected annotations\n",
    "- Quality control (determining outliers by counts as well as mitochondrial genes)\n",
    "- Normalization using analytic Pearson residuals\n",
    "- Feature selection\n",
    "- Dimenstionality reduction and clustering\n",
    "- UMAP embedding\n",
    "- Manual cell type annotation\n",
    "- Visualization of marker expression\n",
    "\n",
    "Some quick features:\n",
    "- We filter out genes that occur in <3 cells\n",
    "- We filter out cells that have < 2000 genes\n",
    "- We filter out cells that are outside 4 median absolute deviations of distinct gene counts or total gene counts\n",
    "- We filter out mitochondrial genes\n",
    "- Batch covariate is Age\n",
    "- We normalize using analytic Pearson residuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bb9bc4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Johann\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "# Importing necessary modules for ...\n",
    "# ... data processing ...\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import bbknn\n",
    "ll\n",
    "# ... plotting ...\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rc_context\n",
    "from matplotlib.colors import ListedColormap\n",
    "import matplotlib.font_manager as fm\n",
    "import seaborn as sns\n",
    "\n",
    "# ... reading and writing files\n",
    "import sys, os\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eccb8c0c",
   "metadata": {},
   "source": [
    "### Plotting Configuration\n",
    "\n",
    "This cell is simply for making pretty plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62d13224",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining custom color map for plots\n",
    "with open(r\".\\Resources\\ColorMap.csv\", \"r\") as file:\n",
    "    my_cmap_df = pd.read_csv(file, header=0)\n",
    "    my_cmap = ListedColormap(my_cmap_df[\"HEX_Code\"].to_list())\n",
    "\n",
    "# Defining save location for plots\n",
    "figures_folder = r\".\\Plots\\\\\"\n",
    "\n",
    "# Defining folder where raw data is stored\n",
    "data_folder = r\".\\Data\\\\\"\n",
    "\n",
    "# Adding custom font\n",
    "fonts= [r\"Resources\\Arimo\\Arimo-Regular.ttf\",\n",
    "        r\".\\Resources\\Arimo\\Arimo-SemiBold.ttf\",\n",
    "        r\".\\Resources\\Arimo\\Arimo-Italic.ttf\",\n",
    "        r\".\\Resources\\Arimo\\Arimo-SemiBoldItalic.ttf\"]\n",
    "for font in fonts:\n",
    "    fm.fontManager.addfont(font)\n",
    "\n",
    "# Setting parameters for ScanPy plots\n",
    "sc.set_figure_params(\n",
    "    dpi= 150,\n",
    "    dpi_save= 450,\n",
    "    facecolor= \"white\",\n",
    "    color_map= \"RdPu\")\n",
    "\n",
    "scale = 1.4 #plot scale factor\n",
    "fig_width = 2.35*scale\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arimo\"\n",
    "plt.rcParams[\"figure.facecolor\"] = (1.0, 1.0, 1.0, 0.0)\n",
    "plt.rcParams[\"axes.labelsize\"] = 6*scale\n",
    "plt.rcParams[\"axes.labelweight\"] = \"semibold\"\n",
    "plt.rcParams[\"axes.facecolor\"] = (1.0, 1.0, 1.0, 0.7)\n",
    "plt.rcParams[\"xtick.labelsize\"] = 6*scale\n",
    "plt.rcParams[\"ytick.labelsize\"] = 6*scale\n",
    "plt.rcParams[\"axes.titlesize\"] = 8*scale\n",
    "plt.rcParams[\"axes.titlelocation\"] = \"left\"\n",
    "plt.rcParams[\"axes.titleweight\"] = \"semibold\"\n",
    "plt.rcParams[\"legend.fontsize\"] = 6*scale"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e7d20bf",
   "metadata": {},
   "source": [
    "### Reading Data\n",
    "\n",
    "We read the data from an .h5ad file that contains the raw count matrix with some annotations and metadata. Between all data sets we calculate the intersection of features contained (common genes between them) and limit each data set to only the genes that are present in all of them.\n",
    "\n",
    "Information carried over is:\n",
    "- plate ID (a unique barcode for each 384-well plate sequenced)\n",
    "- age of animals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c052dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading data\n",
    "adatas = {\"adata_3mo\":  sc.read_h5ad(data_folder + r\"CBA_3mo_SGN.h5ad\"),\n",
    "          \"adata_12mo\": sc.read_h5ad(data_folder + r\"CBA_12mo_SGN.h5ad\"),\n",
    "          \"adata_24mo\": sc.read_h5ad(data_folder + r\"CBA_24mo_SGN.h5ad\")}\n",
    "\n",
    "# Getting list of genes in all data sets\n",
    "features_3mo = adatas[\"adata_3mo\"].var_names.to_list()\n",
    "features_12mo = adatas[\"adata_12mo\"].var_names.to_list()\n",
    "features_24mo = adatas[\"adata_24mo\"].var_names.to_list()\n",
    "print(f\"Identified {len(features_3mo)} genes in first data set.\")\n",
    "print(f\"Identified {len(features_12mo)} genes in second data set.\")\n",
    "print(f\"Identified {len(features_24mo)} genes in third data set.\\n\")\n",
    "\n",
    "common_features = np.intersect1d(np.intersect1d(features_3mo, features_12mo), features_24mo).tolist()\n",
    "print(f\"Identified {len(common_features)} common genes between data sets.\", common_features, sep= \"\\n\")\n",
    "\n",
    "adatas[\"adata_3mo\"] = adatas[\"adata_3mo\"][:, adatas[\"adata_3mo\"].var_names.isin(common_features)].copy()\n",
    "adatas[\"adata_12mo\"] = adatas[\"adata_12mo\"][:, adatas[\"adata_12mo\"].var_names.isin(common_features)].copy()\n",
    "adatas[\"adata_24mo\"] = adatas[\"adata_24mo\"][:, adatas[\"adata_24mo\"].var_names.isin(common_features)].copy()\n",
    "\n",
    "print(f\"\\nFiltered dataset 1 has {adatas[\"adata_3mo\"].n_obs} nuclei and {adatas[\"adata_3mo\"].n_vars} genes.\\n\\\n",
    "Filtered dataset 2 has {adatas[\"adata_12mo\"].n_obs} nuclei and {adatas[\"adata_12mo\"].n_vars} genes.\\n\\\n",
    "Filtered dataset 3 has {adatas[\"adata_24mo\"].n_obs} nuclei and {adatas[\"adata_24mo\"].n_vars} genes.\")\n",
    "\n",
    "# Concatenating all age groups\n",
    "adata_original = ad.concat([adatas[\"adata_3mo\"], adatas[\"adata_12mo\"], adatas[\"adata_24mo\"]],\n",
    "                           label= \"age\",\n",
    "                           keys= [\"3 months\", \"12 months\", \"24 months\"])\n",
    "\n",
    "# Making a clean copy ...\n",
    "adata = ad.AnnData(adata_original.X.copy())\n",
    "\n",
    "# # ... and preserving some of the metadata\n",
    "adata.obs[\"plate\"] = adata_original.obs[\"plate\"].values.astype(str)\n",
    "adata.obs[\"age\"] = adata_original.obs[\"age\"].values\n",
    "adata.obs.set_index(adata_original.obs.index, inplace= True)\n",
    "adata.var.set_index(adata_original.var.index, inplace= True)\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c5df443",
   "metadata": {},
   "source": [
    "### Quality Control\n",
    "\n",
    "We calculate quality control metrics. These include:\n",
    "- total gene counts per nucleus\n",
    "- distinct genes per nucleus\n",
    "- percentages of the defined gene sets (mito, ribo, hemo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81b1a8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define mitochondrial, ribosomal, and hemoglobin genes by name\n",
    "# (they are already filtered out in this dataset ...)\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\")\n",
    "adata.var[\"ribo\"] = adata.var_names.str.startswith((\"Rpl\", \"Rps\"))\n",
    "adata.var[\"hb\"] = adata.var_names.str.startswith(\"Hb\")\n",
    "\n",
    "print(f\"{len(adata.var[adata.var[\"mt\"]])} Mitochondrial genes identified:\\n\", adata.var_names[adata.var[\"mt\"]].values, \"\\n\")\n",
    "print(f\"{len(adata.var[adata.var[\"ribo\"]])} Ribosomal genes identified:\\n\", adata.var_names[adata.var[\"ribo\"]].values, \"\\n\")\n",
    "print(f\"{len(adata.var[adata.var[\"hb\"]])} Hemoglobin genes identified:\\n\", adata.var_names[adata.var[\"hb\"]].values)\n",
    "\n",
    "# Calculating QC metrics including the gene subsets\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars= [\"mt\", \"ribo\", \"hb\"],\n",
    "                           inplace= True,\n",
    "                           log1p= True)\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "267c16a4",
   "metadata": {},
   "source": [
    "### Calling Outliers\n",
    "\n",
    "We define outliers nuclei as follows:\n",
    "- total counts outside of 4 median absolute deviations (MADs) OR\n",
    "- distinct counts outside of 4 MADs OR\n",
    "- distinct counts < 2000 genes\n",
    "\n",
    "The latter is a manual hard cut-off that is based on the minimum performance expected from Smart-seq3xpress."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "633aaed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculating MAD for a given QC metric\n",
    "def is_outlier(adata, metric: str, nmads: int):\n",
    "    '''This function calculates whether each element of a list or series is an\n",
    "    outlier based on the number of median absolute deviations (MADs) passed in the\n",
    "    arguments nmads. It returns a list of booleans where each element that was an outlier\n",
    "    in the original list results in a TRUE in the returned list. If you use the returned\n",
    "    list for indexing to EXCLUDE outliers, remember to index with the inverse (~outliers) of the outlier\n",
    "    list returned by this function'''\n",
    "    M = adata.obs[metric]\n",
    "    outlier = ((M < np.median(M) - nmads * st.median_abs_deviation(M)) # outlier below ...\n",
    "               | (np.median(M) + nmads * st.median_abs_deviation(M) < M)) # and outlier above\n",
    "    return outlier\n",
    "\n",
    "# Defining cut-off for mitochondrial percentage\n",
    "max_mito_pct = 0.3\n",
    "\n",
    "# Defining whether a nucleus is an outlier based on total counts and gene counts per cell\n",
    "adata.obs[\"outlier\"] = (is_outlier(adata, \"log1p_total_counts\", 4)\n",
    "                        | is_outlier(adata, \"log1p_n_genes_by_counts\", 4)\n",
    "                        | (adata.obs[\"n_genes_by_counts\"] < 2000)) # Custom hard threshold for minimum gene number per nucleus due to Smart-seq3xpress performance\n",
    "\n",
    "# Running doublet detection and creating a doublet score for every cell\n",
    "sc.pp.scrublet(adata, batch_key= \"age\")\n",
    "\n",
    "print(f\"Identified {len(adata.obs[adata.obs[\"outlier\"]])} out of {len(adata.obs)} nuclei as outliers.\\n\",\n",
    "      \"Nuclei remaining of each age group: \\n\",\n",
    "      adata[~adata.obs[\"outlier\"]].obs[\"age\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac023410",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Quality Control metrics\n",
    "fig, axes = plt.subplots(3, 4, figsize=(fig_width*4.2, fig_width*3), layout= \"constrained\")\n",
    "\n",
    "age_colors = [my_cmap.colors[4],\n",
    "              my_cmap.colors[6],\n",
    "              my_cmap.colors[5]]\n",
    "# Gene per Nuclei (all samples)\n",
    "sc.pl.violin(adata, [\"n_genes_by_counts\"], jitter= 0.4, ax= axes[1, 0], show= False, palette = my_cmap.colors)\n",
    "axes[1, 0].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[1, 0].set_ylabel(\"Number of Distinct Genes\")\n",
    "axes[1, 0].axhline(adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].min(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 0].axhline(adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].max(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 0].text(-0.4, adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].max()+170, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Genes per Nuclei (grouped by age)\n",
    "sc.pl.violin(adata, [\"n_genes_by_counts\"], jitter= 0.4, ax= axes[1, 2], show= False, palette = age_colors, groupby= \"age\")\n",
    "axes[1, 2].set_xlabel(\"Single Nuclei\")\n",
    "axes[1, 2].set_ylabel(\"Number of Distinct Genes\")\n",
    "axes[1, 2].axhline(adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].min(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 2].axhline(adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].max(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 2].text(-0.4, adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].max()+170, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Total Counts (all samples)\n",
    "sc.pl.violin(adata, [\"total_counts\"], jitter= 0.4, ax= axes[1, 1], show= False, palette = my_cmap.colors)\n",
    "axes[1, 1].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[1, 1].set_ylabel(\"Total UMI Counts\")\n",
    "axes[1, 1].axhline(adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].min(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 1].axhline(adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].max(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 1].text(-0.4, adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].max()+2000, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "axes[1, 1].set_ylim(0, 100000)\n",
    "axes[1, 1].text(-0.4, 97000, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Total Counts (grouped by age)\n",
    "sc.pl.violin(adata, [\"total_counts\"], jitter= 0.4, ax= axes[1, 3], show= False, palette = age_colors, groupby= \"age\")\n",
    "axes[1, 3].set_xlabel(\"Single Nuclei\")\n",
    "axes[1, 3].set_ylabel(\"Total UMI Counts\")\n",
    "axes[1, 3].axhline(adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].min(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 3].axhline(adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].max(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 3].text(-0.4, adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].max()+2000, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "axes[1, 3].set_ylim(0, 100000)\n",
    "axes[1, 3].text(-0.4, 97000, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Mitochondrial percentage (all samples)\n",
    "sc.pl.violin(adata, [\"pct_counts_mt\"], jitter= 0.4, ax= axes[2, 0], show= False, palette = my_cmap.colors)\n",
    "axes[2, 0].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[2, 0].set_ylabel(\"Percentage of Mitochondrial Genes\")\n",
    "axes[2, 0].set_ylim(0, 1)\n",
    "axes[2, 0].text(-0.4, 0.95, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "axes[2, 0].axhline(max_mito_pct,\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[2, 0].text(-0.4, max_mito_pct+max_mito_pct*0.1, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Mitochondrial percentage (grouped by age)\n",
    "sc.pl.violin(adata, [\"pct_counts_mt\"], jitter= 0.4, ax= axes[2, 2], show= False, palette = age_colors, groupby= \"age\")\n",
    "axes[2, 2].set_xlabel(\"Single Nuclei\")\n",
    "axes[2, 2].set_ylabel(\"Percentage of Mitochondrial Genes\")\n",
    "axes[2, 2].set_ylim(0, 1)\n",
    "axes[2, 2].text(-0.4, 0.95, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "axes[2, 2].axhline(max_mito_pct,\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[2, 2].text(-0.4, max_mito_pct+max_mito_pct*0.1, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Ribosomal percentage (all samples)\n",
    "sc.pl.violin(adata, [\"pct_counts_ribo\"], jitter= 0.4, ax= axes[2, 1], show= False, palette = my_cmap.colors)\n",
    "axes[2, 1].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[2, 1].set_ylabel(\"Percentages of Ribosomal Genes\")\n",
    "axes[2, 1].set_ylim(0, 5)\n",
    "axes[2, 1].text(-0.4, 4.8, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Ribosomal percentage (grouped by age)\n",
    "sc.pl.violin(adata, [\"pct_counts_ribo\"], jitter= 0.4, ax= axes[2, 3], show= False, palette = age_colors, groupby= \"age\")\n",
    "axes[2, 3].set_xlabel(\"Single Nuclei\")\n",
    "axes[2, 3].set_ylabel(\"Percentages of Ribosomal Genes\")\n",
    "axes[2, 3].set_ylim(0, 5)\n",
    "axes[2, 3].text(-0.4, 4.8, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Highest expressed genes (all samples)\n",
    "sc.pl.highest_expr_genes(adata, n_top=20, ax= axes[0, 1], show= False, palette= my_cmap.colors)\n",
    "axes[0, 1].set_title(\"Highest Expressed Genes Total\")\n",
    "axes[0, 1].set_xlabel(r\"% of Total Counts\")\n",
    "axes[0, 1].set_xlim(0, 10)\n",
    "axes[0, 1].text(9.6, 19, \"x-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                rotation= 90,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Total Count Distribution (all samples)\n",
    "sns.histplot(adata.obs[\"total_counts\"], bins = 100, kde= False, ax= axes[0, 0], color= my_cmap.colors[0])\n",
    "axes[0, 0].set_title(\"Distribution of Total Counts\")\n",
    "axes[0, 0].set_xlabel(\"Total UMI Counts\")\n",
    "axes[0, 0].set_ylabel(\"Number of Nuclei\")\n",
    "axes[0, 0].set_xlim(0, 100000)\n",
    "axes[0, 0].text(95000, 150, \"x-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                rotation= 90,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Total Counts Distribution (grouped by age)\n",
    "sns.histplot(adata.obs, x= \"total_counts\", bins = 100, kde= False, ax= axes[0, 2], palette= age_colors, hue= \"age\", hue_order= (\"3 months\", \"12 months\", \"24 months\"))\n",
    "axes[0, 2].set_title(\"Distribution of Total Counts\")\n",
    "axes[0, 2].set_xlabel(\"Total UMI Counts\")\n",
    "axes[0, 2].set_ylabel(\"Number of Nuclei\")\n",
    "axes[0, 2].legend([\"3 months\", \"12 months\", \"24 months\"], title= \"Age Group\", title_fontsize= 6*scale, frameon= False)\n",
    "axes[0, 2].set_xlim(0, 100000)\n",
    "axes[0, 2].text(95000, 150, \"x-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                rotation= 90,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "fig.tight_layout(h_pad= 1.0*scale)\n",
    "fig.savefig(figures_folder + r\"QC_Metrics_SGNAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "798da208",
   "metadata": {},
   "source": [
    "### Normalization\n",
    "\n",
    "We normalize counts using analytical pearson residuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ecf2b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving a copy of the raw counts before normalising\n",
    "adata.layers[\"raw_counts\"] = adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85caab53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize the data and store the normalizations in separate layers ...\n",
    "# ... by pearson residulals ...\n",
    "adata.layers[\"pearson_normalized_counts\"] = sc.experimental.pp.normalize_pearson_residuals(adata, copy= True).X\n",
    "\n",
    "# ... and by median (with log1p transformation)\n",
    "adata.layers[\"log1p_normalized_counts\"] = sc.pp.normalize_total(adata, copy= True).X\n",
    "sc.pp.log1p(adata, layer= \"log1p_normalized_counts\")\n",
    "\n",
    "fig, axes = plt.subplots(1, 3, figsize=(fig_width*2.4, fig_width*0.8), layout= \"constrained\")\n",
    "\n",
    "# the original layer remains raw\n",
    "sns.histplot(adata.X.sum(1), bins = 100, kde= False, ax= axes[0], palette= my_cmap.colors, legend= False)\n",
    "axes[0].set_title(\"Non-normalized\")\n",
    "\n",
    "sns.histplot(adata.layers[\"pearson_normalized_counts\"].sum(1), bins = 100, kde= False, ax= axes[1], color= my_cmap.colors[0])\n",
    "axes[1].set_title(\"Analytical Pearson Residuals\")\n",
    "\n",
    "\n",
    "sns.histplot(adata.layers[\"log1p_normalized_counts\"].sum(1), bins = 100, kde= False, ax= axes[2], palette= my_cmap.colors, legend= False)\n",
    "axes[2].set_title(\"Median + Log1p\")\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(figures_folder + r\"Normalization_SGNAtlas.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b404d8bd",
   "metadata": {},
   "source": [
    "### Feature Selection\n",
    "\n",
    "We select the top 2000 most highly variable genes (HVGs) based on pearson normalized and median normalized data and save the annotation so separate fields.\n",
    "\n",
    "**DO NOT RUN THE FOLLOWING CELL TWICE WITHOUT RELOADING THE DATA OR RESTARTING THE KERNEL**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad588393",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Feature selection based on pearson residuals (uses the raw counts by default) ...\n",
    "sc.experimental.pp.highly_variable_genes(adata, n_top_genes= 2000, batch_key= \"age\")\n",
    "\n",
    "# ... and based on log1p normalized counts\n",
    "if \"log1p_highly_variable\" in adata.obs_names.values:\n",
    "    pass\n",
    "else:\n",
    "    adata.var = pd.concat((adata.var,\n",
    "                        sc.pp.highly_variable_genes(adata, n_top_genes= 2000,\n",
    "                                                    layer= \"log1p_normalized_counts\",\n",
    "                                                    inplace= False).rename(\n",
    "                                                        columns={\"highly_variable\": \"log1p_highly_variable\",\n",
    "                                                                    \"means\": \"log1p_means\",\n",
    "                                                                    \"dispersions\": \"log1p_dispersions\",\n",
    "                                                                    \"dispersions_norm\": \"log1p_dispersions_norm\"}).loc[:, [\"log1p_highly_variable\", \"log1p_means\", \"log1p_dispersions\", \"log1p_dispersions_norm\"]]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d856d249",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting highly variable gene fraction\n",
    "fig, ax = plt.subplots(1,2, figsize=(fig_width*1.6, fig_width*0.8), layout= \"constrained\")\n",
    "\n",
    "sns.scatterplot(adata.var, x= \"means\", y= \"residual_variances\",\n",
    "                hue= \"highly_variable\", \n",
    "                palette= my_cmap.colors,\n",
    "                ax= ax[0], \n",
    "                s= 5,\n",
    "                edgecolor= \"black\",\n",
    "                linewidth= 0.1,\n",
    "                zorder= 2)\n",
    "ax[0].set_title(\"Highly Variable Genes Pearson\")\n",
    "ax[0].legend()\n",
    "ax[0].set_xscale(\"log\")\n",
    "ax[0].set_yscale(\"log\")\n",
    "\n",
    "sns.scatterplot(adata.var, x= \"log1p_means\", y= \"log1p_dispersions\",\n",
    "                hue= \"log1p_highly_variable\",\n",
    "                palette= my_cmap.colors,\n",
    "                ax= ax[1],\n",
    "                s= 5,\n",
    "                edgecolor= \"black\",\n",
    "                linewidth= 0.1,\n",
    "                zorder= 2)\n",
    "ax[1].set_title(\"Highly Variable Genes Log1p\")\n",
    "ax[1].set_xscale(\"log\")\n",
    "ax[1].legend()\n",
    "\n",
    "fig.savefig(figures_folder + r\"HVGs_SGNAtlas.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f535108",
   "metadata": {},
   "source": [
    "### Dimensionality Reduction\n",
    "\n",
    "We compute principal components on the pearson normalized count matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82b58f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making sure raw counts are used\n",
    "adata.X = adata.layers[\"raw_counts\"].copy()\n",
    "\n",
    "# Computing PCA with pearson residuals on the raw count matrix\n",
    "sc.experimental.pp.normalize_pearson_residuals_pca(adata)\n",
    "\n",
    "# We inspect the PCA results\n",
    "fig, axes = plt.subplots(2, 3, figsize=(fig_width*3.5, fig_width*1.6), layout= \"compressed\")\n",
    "\n",
    "sc.pl.pca_variance_ratio(adata, show= False)\n",
    "axes[0, 0].set_title(\"Variance Ratios\")\n",
    "\n",
    "# PCA by plate\n",
    "sc.pl.pca(adata, color= \"plate\",\n",
    "          palette= my_cmap.colors,\n",
    "          s= 10,\n",
    "          edgecolor= \"black\",\n",
    "          linewidth= 0.1,\n",
    "          ax= axes[0, 1],\n",
    "          sort_order= False,\n",
    "          show= False)\n",
    "axes[0, 1].set_title(\"PCA Results (by Plate ID)\")\n",
    "\n",
    "# PCA by total counts\n",
    "sc.pl.pca(adata, color= \"total_counts\",\n",
    "          s= 10,\n",
    "          edgecolor= \"black\",\n",
    "          linewidth= 0.1,\n",
    "          ax= axes[0, 2],\n",
    "          sort_order= False,\n",
    "          show= False)\n",
    "axes[0, 2].set_title(\"PCA Results by Total Counts\")\n",
    "\n",
    "# PCA by age group\n",
    "sc.pl.pca(adata, color= \"age\",\n",
    "          palette= my_cmap.colors,\n",
    "          s= 10,\n",
    "          edgecolor= \"black\",\n",
    "          linewidth= 0.1,\n",
    "          ax= axes[1, 0],\n",
    "          sort_order= False,\n",
    "          show= False)\n",
    "axes[1, 0].set_title(\"PCA Results (by age group)\")\n",
    "\n",
    "fig.savefig(figures_folder + r\"PCA_SGNAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cba51de",
   "metadata": {},
   "source": [
    "### UMAP Embedding\n",
    "\n",
    "We batch correct the data by calculating a batch-balances k-nearest neighbor (BBKNN) graph and calculate a UMAP projection.\n",
    "The batch covariate is the plateID, so we correct for differences between each plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eccb3c02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making sure normalized counts are used\n",
    "adata.X = adata.layers[\"pearson_normalized_counts\"].copy()\n",
    "\n",
    "# BBKNN embedding\n",
    "bbknn.bbknn(adata, batch_key= \"age\", neighbors_within_batch= 3)\n",
    "\n",
    "# UMAP Embedding\n",
    "sc.tl.umap(adata, min_dist= 1.3, spread= 1.6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b58c532",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting before filtering\n",
    "fig, axs = plt.subplots(2, 3, figsize= (fig_width*3.5, fig_width*1.6), layout= \"compressed\")\n",
    "\n",
    "metrics = [\"total_counts\", \"age\", \"predicted_doublet\", \"n_genes_by_counts\", \"age\", \"plate\"]\n",
    "dot_size = 6\n",
    "vmax = \"p99\"\n",
    "for i, metric in enumerate(metrics):\n",
    "    if metric == \"age\":\n",
    "        cmap = [my_cmap.colors[4],\n",
    "                my_cmap.colors[6],\n",
    "                my_cmap.colors[5]]\n",
    "    else:\n",
    "        cmap = my_cmap.colors\n",
    "    sc.pl.umap(adata,\n",
    "            color=[metric],\n",
    "            palette= cmap,\n",
    "            size= dot_size,\n",
    "            vmax= vmax,\n",
    "            edgecolor= \"black\",\n",
    "            linewidth= 0.1,\n",
    "            ax= axs[int(i/(len(metrics)/2)), int(i%(len(metrics)/2))],\n",
    "            show= False)\n",
    "axs[0, 0].text(-16,27, f\"n = {adata.n_obs}\", size= 8)\n",
    "\n",
    "fig.savefig(figures_folder + r\"UMAP_Prefilter_SGNAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c2482e3",
   "metadata": {},
   "source": [
    "### Filtering\n",
    "\n",
    "We filter out cells and genes that we determined to be outliers as defined above.<br>\n",
    "Exclusion criteria:\n",
    "- total genes counts outside of 4 MADs in positive or negative direction\n",
    "- less than 1000 genes expressed per nucleus (due to Smart-seq3xpress performance)\n",
    "- single genes expressed in less than 3 cells in data set\n",
    "- all mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eeda111",
   "metadata": {},
   "outputs": [],
   "source": [
    "# And this is also what we have to do with out custom \"outlier\" annotations\n",
    "print(f\"Number of nuclei before outliers: {adata.n_obs}\\n\\\n",
    "Number of genes before outliers: {adata.n_vars}\")\n",
    "adata = adata[(~adata.obs[\"outlier\"]), ~adata.var[\"mt\"]]\n",
    "print(f\"\\nNumber of nuclei after outliers: {adata.n_obs}\\n\\\n",
    "Number of genes after outliers: {adata.n_vars}\")\n",
    "\n",
    "print(f\"\\nNumber of nuclei before count exclusion: {len(adata.obs)}\")\n",
    "sc.pp.filter_genes(adata, min_cells= 3, inplace= True)\n",
    "print(f\"Number of nuclei after count exclusion: {len(adata.obs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a13606a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting before filtering\n",
    "fig, axs = plt.subplots(2, 3, figsize= (fig_width*3.5, fig_width*1.6), layout= \"compressed\")\n",
    "\n",
    "metrics = [\"total_counts\", \"age\", \"predicted_doublet\", \"n_genes_by_counts\", \"age\", \"plate\"]\n",
    "dot_size = 6\n",
    "vmax = \"p99\"\n",
    "for i, metric in enumerate(metrics):\n",
    "    if metric == \"age\":\n",
    "        cmap = [my_cmap.colors[4],\n",
    "                my_cmap.colors[6],\n",
    "                my_cmap.colors[5]]\n",
    "    else:\n",
    "        cmap = my_cmap.colors\n",
    "    sc.pl.umap(adata,\n",
    "            color=[metric],\n",
    "            palette= cmap,\n",
    "            size= dot_size,\n",
    "            vmax= vmax,\n",
    "            edgecolor= \"black\",\n",
    "            linewidth= 0.1,\n",
    "            ax= axs[int(i/(len(metrics)/2)), int(i%(len(metrics)/2))],\n",
    "            show= False)\n",
    "axs[0, 0].text(-16,27, f\"n = {adata.n_obs}\", size= 8)\n",
    "\n",
    "fig.savefig(figures_folder + r\"UMAP_Postfilter_SGNAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e36980d",
   "metadata": {},
   "source": [
    "### Clustering and Annotating\n",
    "\n",
    "We cluster using the Leiden algorithm at an initial resolution of 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbd19f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers[\"pearson_normalized_counts\"].copy() # The leiden() function requires normalized counts\n",
    "\n",
    "# Clustering ...\n",
    "sc.tl.leiden(adata,\n",
    "             flavor= \"igraph\",\n",
    "             n_iterations= -1, # iterate until the best clusters have been found\n",
    "             resolution= 1, # higher number means finer clusters\n",
    "             key_added= \"leiden\")\n",
    "\n",
    "# Plotting the clusters\n",
    "fig, ax = plt.subplots(1,1, figsize= (fig_width*0.9, fig_width*0.9))\n",
    "\n",
    "sc.pl.umap(adata, color= [\"leiden\"],\n",
    "           palette= my_cmap.colors,\n",
    "           size= 6,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1,\n",
    "           legend_loc= \"on data\",\n",
    "           legend_fontsize = 3,\n",
    "           legend_fontoutline = 0.4,\n",
    "           title= \"Leiden Clustering (resolution 1)\",\n",
    "           ax= ax,\n",
    "           show= False)\n",
    "\n",
    "fig.savefig(figures_folder + r\"Leiden_SGNAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "967506ee",
   "metadata": {},
   "source": [
    "#### Loading Marker Genes\n",
    "\n",
    "A list of literature markers of different cochlear cell types. The same ones as used for the reference atlas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27bd8dd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Marker Genes\n",
    "# Loading previously saved markers\n",
    "with open(r\"Resources\\cochlea_markers.json\", \"r\") as file:\n",
    "    markers_SGN = json.load(file)\n",
    "\n",
    "# Checking which marker genes are present in the dataset\n",
    "markers_in_data = {}\n",
    "for ct, markers in markers_SGN.items():\n",
    "    if np.intersect1d(markers, adata.var_names).size > 0:\n",
    "        markers_in_data[ct] = list(np.intersect1d(markers, adata.var_names))\n",
    "\n",
    "# printing the final list of available marker genes\n",
    "print(\"Marker genes present in data:\\n\")\n",
    "for item in markers_in_data.items():\n",
    "    print(f\"{item[0]}: {item[1]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b04e215",
   "metadata": {},
   "source": [
    "### Checking Marker Expression\n",
    "\n",
    "We visualize the expression of the marker genes in each cluster and infer which clusters could belong to which cell type.\n",
    "\n",
    "Initally we just observe if the markers locate to clusters unambiguously, if not we sub-cluster the clusters that are too coarse in the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf2e266",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Next, we check the expression patterns of the marker genes in our clustering\n",
    "sc.tl.leiden(adata,\n",
    "             flavor= \"igraph\",\n",
    "             n_iterations= -1,\n",
    "             resolution= 0.8,\n",
    "             key_added= \"leiden_0\")\n",
    "\n",
    "# Dotplot featuring all marker genes and all the clusters to manually determine cell identities in clusters\n",
    "# THIS IS WHERE THE ACTUAL CLUSTER ANNOTATION CALLING HAPPENS\n",
    "sc.pl.dotplot(adata,\n",
    "              markers_in_data,\n",
    "              groupby=\"leiden_0\",\n",
    "              standard_scale=\"var\",\n",
    "              cmap= \"Purples\", # I like purple\n",
    "            #   save= \"Marker_Genes_SGNAtlas.png\", # save the dotplot, comment out if desired\n",
    "              )\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"leiden_0\"],\n",
    "           legend_loc=\"on data\",\n",
    "           legend_fontsize= 2,\n",
    "           legend_fontoutline= 0.8,\n",
    "           size= 6,\n",
    "           palette= my_cmap.colors,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d78602ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# After initial inspection of clustering, we defining clusters that look like they wanna be clustered further\n",
    "# This is iterative sub-clustering\n",
    "need_subclustering = {\"1\": 0.5,\n",
    "                      \"10\": 0.5,\n",
    "                      \"11\": 0.2}\n",
    "\n",
    "# Successive sub-clustering of clusters that are too coarse to tell cell types apart\n",
    "i = 1\n",
    "for cluster, res in need_subclustering.items():\n",
    "    sc.tl.leiden(adata, flavor= \"igraph\", n_iterations= -1, key_added= f\"leiden_{i}\",\n",
    "                 resolution= res,\n",
    "                 restrict_to= (f\"leiden_{i-1}\", [cluster]))\n",
    "    i += 1\n",
    "\n",
    "# Dotplot featuring all marker genes and all the clusters to manually determine cell identities in clusters\n",
    "# THIS IS WHERE THE ACTUAL CLUSTER ANNOTATION CALLING HAPPENS\n",
    "sc.pl.dotplot(adata,\n",
    "              markers_in_data,\n",
    "              groupby=f\"leiden_{len(need_subclustering)}\",\n",
    "              standard_scale= \"var\",\n",
    "              cmap= \"Purples\", # I like purple\n",
    "            #   save= \"Marker_Genes_SGNAtlas.png\", # Comment in if wanted\n",
    "              )\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[f\"leiden_{len(need_subclustering)}\"],\n",
    "           legend_loc=\"on data\",\n",
    "           legend_fontsize= 2,\n",
    "           legend_fontoutline= 0.8,\n",
    "           size= 6,\n",
    "           palette= my_cmap.colors,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb4eb660",
   "metadata": {},
   "source": [
    "### Checking Cluster Expressions\n",
    "\n",
    "We visualise marker gene expressions on the UMAP projection to visually verify localizations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8553623d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining a subgroup of cell types of which to check expression levels across the whole data set\n",
    "# Change cell types as needed!\n",
    "sgns = [\n",
    "    \"Spiral Ganglion Neuron Type Ia\",\n",
    "    \"Spiral Ganglion Neuron Type Ib\",\n",
    "    \"Spiral Ganglion Neuron Type Ic\",\n",
    "    \"Spiral Ganglion Neuron Type II\"]\n",
    "\n",
    "# Now we plot marker gene expression for each marker of the above defined cell types\n",
    "for ct in sgns:\n",
    "    print(f\"{ct}:\")  # print cell subtype name\n",
    "    sc.pl.umap(adata,\n",
    "               color= [*markers_in_data[ct], f\"leiden_{len(need_subclustering)}\"],\n",
    "               vmax=\"p99\",\n",
    "               sort_order=False,  # do not plot highest expression on top, to not get a biased view of the mean expression among cell,\n",
    "               frameon=False,\n",
    "               size= 6,\n",
    "               palette= my_cmap.colors,\n",
    "               edgecolor= \"black\",\n",
    "               linewidth= 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01cae918",
   "metadata": {},
   "source": [
    "### Comitting Annotations\n",
    "\n",
    "We now define a manual cell type for each numeric cluster from before and map the annotated cells to a new field in the .obs annotations (every nucleus gets a cell type label)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a49cd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Committing Annotations\n",
    "annotations = {\"0\": \"Myelinating Schwann Cells\",\n",
    "               \"1,0\": \"Fibrocytes I\",\n",
    "               \"1,1\": \"Fibrocytes I\",\n",
    "               \"1,2\": \"Fibrocytes II\",\n",
    "               \"1,3\": \"Fibrocytes IV\",\n",
    "               \"1,4\": \"Unclear\",\n",
    "               \"1,5\": \"Unclear\",\n",
    "               \"2\": \"Satellite Glial Cells 1\",\n",
    "               \"3\": \"Myelinating Schwann Cells\",\n",
    "               \"4\": \"Myelinating Schwann Cells\",\n",
    "               \"5\": \"Endothelial Cells\",\n",
    "               \"6\": \"Satellite Glial Cells 2\",\n",
    "               \"7\": \"Fibroblasts\",\n",
    "               \"8\": \"Osteocytes\",\n",
    "               \"9\": \"Chondrocytes\",\n",
    "               \"10,0\": \"Spiral Ganglion Neuron Type Ib\",\n",
    "               \"10,2\": \"Spiral Ganglion Neuron Type Ib\",\n",
    "               \"10,1\": \"Spiral Ganglion Neuron Type Ic\",\n",
    "               \"10,3\": \"Spiral Ganglion Neuron Type Ic\",\n",
    "               \"11,0\": \"Spiral Ganglion Neuron Type II\",\n",
    "               \"11,1\": \"Unclear\",\n",
    "               \"11,2\": \"Unclear\",\n",
    "               \"11,3\": \"Unclear\",\n",
    "               \"12\": \"Fibrocyte III\",\n",
    "               \"13\": \"Spiral Ganglion Neuron Type II\",\n",
    "               \"14\": \"Oligodendrocytes\",\n",
    "               \"15\": \"Spiral Ganglion Neuron Type Ia\",\n",
    "               \"16\": \"Satellite Glial Cells 1\",\n",
    "               \"17\": \"Myelinating Schwann Cells\",\n",
    "               \"18\": \"Macrophages\",\n",
    "               \"19\": \"Non-myelinating Schwann Cells\",\n",
    "               \"20\": \"Unclear\",\n",
    "               \"21\": \"Unclear\",\n",
    "               \"22\": \"Myelinating Schwann Cells\",\n",
    "               \"23\": \"Myelinating Schwann Cells\",\n",
    "               \"24\": \"Myelinating Schwann Cells\",\n",
    "               \"25\": \"Myelinating Schwann Cells\",\n",
    "               \"26\": \"Myelinating Schwann Cells\",\n",
    "               \"27\": \"Myelinating Schwann Cells\",\n",
    "               \"28\": \"Myelinating Schwann Cells\"}\n",
    "\n",
    "# Refactoring the cluster identity and saving it to a new annotation as cell types\n",
    "adata.obs[\"manual_celltypes\"] = adata.obs[f\"leiden_{len(need_subclustering)}\"].map(annotations)\n",
    "\n",
    "# Saving the annotated dataset\n",
    "adata.write_h5ad(data_folder + r\"CBA_SGNAtlas_Annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ba9d056",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Cell Type Annotations\n",
    "fig, ax = plt.subplots(2,1, figsize=(fig_width*2.1, fig_width*1.8), layout= \"compressed\")\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"manual_celltypes\"],\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.15,\n",
    "           size= 10,\n",
    "           palette= my_cmap.colors,\n",
    "           legend_fontsize= 6*scale,\n",
    "           ax = ax[0],\n",
    "           show= False)\n",
    "ax[0].set_title(\"SGN Aging Atlas\")\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"age\"],\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.15,\n",
    "           size= 10,\n",
    "           palette= [my_cmap.colors[4],\n",
    "                     my_cmap.colors[6],\n",
    "                     my_cmap.colors[5]],\n",
    "           legend_fontsize= 6*scale,\n",
    "           ax = ax[1],\n",
    "           show= False)\n",
    "ax[1].set_title(\"Age Difference\", fontsize= 10)\n",
    "\n",
    "fig.savefig(figures_folder + r\"Annotations_SGNAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a28bd5",
   "metadata": {},
   "source": [
    "## Gene Expression Analysis\n",
    "\n",
    "Here, as an example, we filter out only the spiral ganglion neurons and plot the expression of one selected marker for each cluster on top of the UMAP projection.<br>\n",
    "\n",
    "For this we reload the annotated file as to not have to rerun all the above cells when reanalysing the data.\n",
    "\n",
    "Further, I am performing a rudimentary **Differential Gene Expression Analysis** using the `sc.tl.rank_genes_groups()` method that determines differentially expressed genes between defined groups and ranks them by giving each gene a differential expression score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e512b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading previously annotated data\n",
    "adata = sc.read_h5ad(data_folder + r\"CBA_SGNAtlas_Annotated.h5ad\")\n",
    "\n",
    "# And reducing the dataset to only SGNs\n",
    "adata_SGNonly = adata[adata.obs[\"manual_celltypes\"].isin([\"Spiral Ganglion Neuron Type II\",\n",
    "                                                          \"Spiral Ganglion Neuron Type Ia\",\n",
    "                                                          \"Spiral Ganglion Neuron Type Ib\",\n",
    "                                                          \"Spiral Ganglion Neuron Type Ic\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a091efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_to_plot = [\"Gata3\", # Type II\n",
    "                   \"Calb2\", # Type Ia\n",
    "                   \"Ttn\",   # Type Ib\n",
    "                   \"Lypd1\"] # Type Ic\n",
    "\n",
    "# Plotting\n",
    "fig, axes = plt.subplots(2,3, figsize=(fig_width*3.4, fig_width*1.6), layout= \"compressed\")\n",
    "\n",
    "dot_size = 50\n",
    "line_width = 0.3\n",
    "vmax = \"p99\"\n",
    "sc.pl.umap(adata_SGNonly,\n",
    "           color=[\"manual_celltypes\"],\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= line_width,\n",
    "           size= dot_size,\n",
    "           palette= my_cmap.colors[14:18],\n",
    "           ax= axes[0, 0],\n",
    "           show= False,\n",
    "           vmax= vmax,\n",
    "           title= \"\")\n",
    "\n",
    "sc.pl.umap(adata_SGNonly[adata_SGNonly.obs[\"manual_celltypes\"].isin([\"Spiral Ganglion Neuron Type Ib\", \"Spiral Ganglion Neuron Type Ic\"])],\n",
    "           color=[\"age\"],\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= line_width,\n",
    "           size= dot_size*2.3,\n",
    "           palette= my_cmap.colors[4:],\n",
    "           ax= axes[1, 0],\n",
    "           show= False,\n",
    "           vmax= vmax,\n",
    "           title= \"Age Difference\")\n",
    "\n",
    "for i, mark in enumerate(markers_to_plot):\n",
    "    sc.pl.umap(adata_SGNonly,\n",
    "               color= mark,\n",
    "               size= dot_size,\n",
    "               vmax= vmax,\n",
    "               legend_loc= \"none\",\n",
    "               edgecolor= \"black\",\n",
    "               linewidth= line_width,\n",
    "               ax= axes[int(i/2), (i%2)+1],\n",
    "               title= mark,\n",
    "               show= False)\n",
    "    axes[int(i/2), (i%2)+1].set_ylabel(\"Normalized Expression\")\n",
    "\n",
    "fig.savefig(figures_folder + r\"SGN_MarkerExpression_SGNplates.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e83c93d3",
   "metadata": {},
   "source": [
    "Differential Gene Expression between age groups. Here we calculate the gene ranks for only Type Ic between age groups or only Type Ib between age groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "636002ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which cell type to analyse\n",
    "cell_type = \"Type Ic\"\n",
    "\n",
    "# slicing the data frame to contain only that one cell type\n",
    "adata_DEG = adata_SGNonly[adata_SGNonly.obs[\"manual_celltypes\"].str.contains(cell_type)].copy()\n",
    "\n",
    "# Ranking differential gene groups in between ages (we compare 24mo to 3mo)\n",
    "sc.tl.rank_genes_groups(adata_DEG, \"age\",\n",
    "                        group= [\"24 months\"],\n",
    "                        reference= \"3 months\",\n",
    "                        method='wilcoxon',\n",
    "                        n_genes= 10)\n",
    "\n",
    "# Plotting\n",
    "fig, axs = plt.subplots(1,2, figsize= (fig_width*2.4, fig_width*1.1), layout= \"compressed\", width_ratios= (1,2))\n",
    "\n",
    "sc.pl.rank_genes_groups(adata_DEG, n_genes=10, ax= axs[1], show= False,\n",
    "                        title= \"DEGs 24 months vs. 3 months\")\n",
    "axs[1].set_axis_off()\n",
    "axs[1].set_ylim(2,10)\n",
    "\n",
    "sc.pl.rank_genes_groups_violin(adata_DEG, groups= [\"24 months\"],\n",
    "                               gene_names= [\"Prl\", \"Gh\"],\n",
    "                               ax= axs[0],\n",
    "                               size= 2.5,\n",
    "                               jitter= 0.25)\n",
    "axs[0].set_title(f\"SGN {cell_type}\")\n",
    "axs[0].set_zorder(3)\n",
    "\n",
    "fig.savefig(figures_folder + f\"DEGs_{cell_type}.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94454c88",
   "metadata": {},
   "source": [
    "### Gene Expression Comparison\n",
    "\n",
    "As supporting data to my cell stains I compare the expressions of the stained markers between young and old."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957f1e9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a dataframe with the plotting data\n",
    "data = dict()\n",
    "adata = sc.read_h5ad(data_folder + r\"CBA_SGNAtlas_Annotated.h5ad\")\n",
    "\n",
    "# For this comparison we use the median + log1p normalized counts because they start at 0\n",
    "adata.X = adata.layers[\"log1p_normalized_counts\"].copy()\n",
    "\n",
    "# Filtering out only Type Ic nuclei\n",
    "# Change this to Type Ia and Ib (list format)\n",
    "adata_SGNonly = adata[adata.obs[\"manual_celltypes\"].isin([\"Spiral Ganglion Neuron Type Ic\"])].copy()\n",
    "\n",
    "# Extracting normalized gene counts for each of the transcripts\n",
    "data[\"3 months\"] = {\"Lypd1\": np.array(adata_SGNonly[adata_SGNonly.obs[\"age\"] == \"3 months\", \"Lypd1\"].to_df()[\"Lypd1\"].tolist()),\n",
    "                    \"Calb2\": np.array(adata_SGNonly[adata_SGNonly.obs[\"age\"] == \"3 months\", \"Calb2\"].to_df()[\"Calb2\"].tolist())}\n",
    "data[\"24 months\"] = {\"Lypd1\": np.array(adata_SGNonly[adata_SGNonly.obs[\"age\"] == \"24 months\", \"Lypd1\"].to_df()[\"Lypd1\"].tolist()),\n",
    "                     \"Calb2\": np.array(adata_SGNonly[adata_SGNonly.obs[\"age\"] == \"24 months\", \"Calb2\"].to_df()[\"Calb2\"].tolist())}\n",
    "\n",
    "\n",
    "# Plotting\n",
    "fig, axs = plt.subplots(1,2, figsize= (fig_width*1.6, fig_width*1), layout= \"tight\", sharey= True)\n",
    "\n",
    "for i, marker in enumerate([\"Lypd1\", \"Calb2\"]):\n",
    "    for j, age in enumerate(data.keys()):\n",
    "        # making a bar plot\n",
    "        axs[i].bar([age],\n",
    "                np.mean(data[age][marker]),\n",
    "                color= my_cmap.colors[j+4],\n",
    "                    edgecolor= \"black\",\n",
    "                    linewidth = 1,\n",
    "                    yerr= np.std(data[age][marker]),\n",
    "                    error_kw= {\"ecolor\": \"black\",\n",
    "                                \"linewidth\": 1.5,\n",
    "                                \"capsize\": 10},\n",
    "                    zorder= 2)\n",
    "        \n",
    "        # Adding jitter points\n",
    "        axs[i].scatter(np.random.normal(j, 0.1, len(data[age][marker])).tolist(),\n",
    "           data[age][marker],\n",
    "           color = \"black\",\n",
    "           s = 6,\n",
    "           zorder= 3)\n",
    "        axs[i].set_title(marker)\n",
    "    else:\n",
    "        # We calculate a wilcoxon Rank sum test to check whether the means are actually not different\n",
    "        print(f\"{marker} Ranksum Test:\\n\", st.ranksums(data[\"3 months\"][marker], data[\"24 months\"][marker]))\n",
    "\n",
    "axs[0].set_ylabel(\"Normalized Expression\")\n",
    "axs[0].set_ylim(-0.2, 4)\n",
    "\n",
    "fig.savefig(figures_folder + r\"Expression_MarkerGenes_Staining_SGNAtlas.png\", dpi= 600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ff2f59",
   "metadata": {},
   "source": [
    "### END OF SCRIPT\n",
    "This notebook was created by *Johann Korn* on 01-10-2025."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
