{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a6eee7f7",
   "metadata": {},
   "source": [
    "## Preprocessing and Analysis of snRNAseq Data\n",
    "\n",
    "In this notebook I do quality control, preprocessing, and data analysis of single nuclei sequencing data (gene count matrices) generated by the Smart-seq3xpress protocol. For this I am mainly using the ScanPy module.<br>\n",
    "**Data used here is of the Whole Cochlea Reference Atlas**\n",
    "\n",
    "The workflow presented here contains:\n",
    "- Reading the data\n",
    "- Removing annotation layers and only reatining some selected annotations\n",
    "- Quality control (determining outliers by counts as well as mitochondrial genes)\n",
    "- Normalization using analytic Pearson residuals\n",
    "- Feature selection\n",
    "- Dimenstionality reduction and clustering\n",
    "- UMAP embedding\n",
    "- Manual cell type annotation\n",
    "- Visualization of marker expression\n",
    "\n",
    "Some quick features:\n",
    "- We filter out genes that occur in <3 cells\n",
    "- We filter out cells that have < 2000 genes\n",
    "- We filter out cells that are outside 4 median absolute deviations of distinct gene counts or total gene counts\n",
    "- We filter out mitochondrial genes (already done in this data set)\n",
    "- Batch covariate is Plate\n",
    "- We normalize using analytic Pearson residuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25a9a4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing necessary modules for ...\n",
    "# ... data processing ...\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import bbknn\n",
    "\n",
    "# ... plotting ...\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rc_context\n",
    "from matplotlib.colors import ListedColormap\n",
    "import matplotlib.font_manager as fm\n",
    "import seaborn as sns\n",
    "\n",
    "# ... reading and writing files\n",
    "import sys, os\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c77b7efd",
   "metadata": {},
   "source": [
    "### Plotting Configuration\n",
    "\n",
    "This cell is simply for making pretty plots and defining some folder paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "441a49f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining custom color map for plots\n",
    "with open(r\".\\Resources\\ColorMap.csv\", \"r\") as file:\n",
    "    my_cmap_df = pd.read_csv(file, header=0)\n",
    "    my_cmap = ListedColormap(my_cmap_df[\"HEX_Code\"].to_list())\n",
    "\n",
    "# Defining save location for plots\n",
    "figures_folder = r\".\\Plots\\\\\"\n",
    "\n",
    "# Defining folder where raw data is stored\n",
    "data_folder = r\".\\Data\\\\\"\n",
    "\n",
    "# Adding custom font\n",
    "fonts= [r\"Resources\\Arimo\\Arimo-Regular.ttf\",\n",
    "        r\".\\Resources\\Arimo\\Arimo-SemiBold.ttf\",\n",
    "        r\".\\Resources\\Arimo\\Arimo-Italic.ttf\",\n",
    "        r\".\\Resources\\Arimo\\Arimo-SemiBoldItalic.ttf\"]\n",
    "for font in fonts:\n",
    "    fm.fontManager.addfont(font)\n",
    "\n",
    "# Setting parameters for ScanPy plots\n",
    "sc.set_figure_params(\n",
    "    dpi= 150,\n",
    "    dpi_save= 450,\n",
    "    facecolor= \"white\",\n",
    "    color_map= \"RdPu\")\n",
    "\n",
    "scale = 1.4 #plot scale factor\n",
    "fig_width = 2.35*scale\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arimo\"\n",
    "plt.rcParams[\"figure.facecolor\"] = (1.0, 1.0, 1.0, 0.0)\n",
    "plt.rcParams[\"axes.labelsize\"] = 6*scale\n",
    "plt.rcParams[\"axes.labelweight\"] = \"semibold\"\n",
    "plt.rcParams[\"axes.facecolor\"] = (1.0, 1.0, 1.0, 0.7)\n",
    "plt.rcParams[\"xtick.labelsize\"] = 6*scale\n",
    "plt.rcParams[\"ytick.labelsize\"] = 6*scale\n",
    "plt.rcParams[\"axes.titlesize\"] = 8*scale\n",
    "plt.rcParams[\"axes.titlelocation\"] = \"left\"\n",
    "plt.rcParams[\"axes.titleweight\"] = \"semibold\"\n",
    "plt.rcParams[\"legend.fontsize\"] = 6*scale"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e7d20bf",
   "metadata": {},
   "source": [
    "### Reading Data\n",
    "\n",
    "We read the data from an .h5ad file that contains the raw count matrix with some annotations and metadata.\n",
    "\n",
    "Information carried over is:\n",
    "- plate ID (a unique barcode for each 384-well plate sequenced)\n",
    "- age of animals\n",
    "- pre-annotated cell types (I make my own, but I use those for verification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c052dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading the data\n",
    "adata_original = sc.read_h5ad(data_folder + r\"CBA_3mo_WholeCochlea.h5ad\")\n",
    "\n",
    "# Making a clean copy ...\n",
    "adata = ad.AnnData(adata_original.layers[\"umi_raw\"].copy()) # Copying over the raw counts\n",
    "\n",
    "# ... and preserving some of the metadata as well as prior cell type annotations\n",
    "adata.obs[\"plate\"] = adata_original.obs[\"plateBC\"].values\n",
    "adata.obs[\"age\"] = \"3 months\"\n",
    "adata.obs[\"cell_types_danny\"] = adata_original.obs[\"cell_types_for_UMAP\"].values\n",
    "adata.obs.set_index(adata_original.obs.index, inplace= True)\n",
    "adata.var.set_index(adata_original.var.index, inplace= True)\n",
    "\n",
    "# Removing potential duplicates\n",
    "adata.obs_names_make_unique()\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba2966a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting SGN plates from other dataset to annotate sample origin\n",
    "adata_SGNonly = sc.read_h5ad(data_folder + r\"CBA_3mo_SGN.h5ad\")\n",
    "\n",
    "total_plates = set(adata.obs[\"plate\"])\n",
    "plates_SGN = set(adata.obs[\"plate\"]).intersection(set(adata_SGNonly.obs[\"plate\"]))\n",
    "print(f\"Identified {len(plates_SGN)} (out of {len(total_plates)}) plates containing SGNs:\\n{plates_SGN}\")\n",
    "\n",
    "adata.obs[\"sample\"] = \"OC\"\n",
    "adata.obs.loc[adata.obs[\"plate\"].isin(plates_SGN), \"sample\"] = \"SGN\"\n",
    "adata.obs[\"sample\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "862cd250",
   "metadata": {},
   "source": [
    "### Quality Control\n",
    "\n",
    "We calculate quality control metrics. These include:\n",
    "- total gene counts per nucleus\n",
    "- distinct genes per nucleus\n",
    "- percentages of the defined gene sets (mito, ribo, hemo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3268667",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define mitochondrial, ribosomal, and hemoglobin genes by name\n",
    "# (mito are already filtered out in this dataset ...)\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\")\n",
    "adata.var[\"ribo\"] = adata.var_names.str.startswith((\"Rpl\", \"Rps\"))\n",
    "adata.var[\"hb\"] = adata.var_names.str.startswith(\"Hb\")\n",
    "\n",
    "print(f\"{len(adata.var_names[adata.var[\"mt\"]])} Mitochondrial genes identified:\\n\", adata.var_names[adata.var[\"mt\"]].values, \"\\n\")\n",
    "print(f\"{len(adata.var_names[adata.var[\"ribo\"]])} Ribosomal genes identified:\\n\", adata.var_names[adata.var[\"ribo\"]].values, \"\\n\")\n",
    "print(f\"{len(adata.var_names[adata.var[\"hb\"]])} Hemoglobin genes identified:\\n\", adata.var_names[adata.var[\"hb\"]].values)\n",
    "\n",
    "# Calculating QC metrics including the gene subsets\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars= [\"mt\", \"ribo\", \"hb\"],\n",
    "                           inplace= True,\n",
    "                           log1p= True)\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19f7d166",
   "metadata": {},
   "source": [
    "### Calling Outliers\n",
    "\n",
    "We define outliers nuclei as follows:\n",
    "- total counts outside of 4 median absolute deviations (MADs) OR\n",
    "- distinct counts outside of 4 MADs OR\n",
    "- distinct counts < 2000 genes\n",
    "\n",
    "The latter is a manual hard cut-off that is based on the minimum performance expected from Smart-seq3xpress."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2329bfb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculating MAD for a given QC metric\n",
    "def is_outlier(adata, metric: str, nmads: int):\n",
    "    '''This function calculates whether each element of a list or series is an\n",
    "    outlier based on the number of median absolute deviations (MADs) passed in the\n",
    "    argument nmads. It returns a list of booleans where each element that was an outlier\n",
    "    in the original list results in a TRUE in the returned list. If you use the returned\n",
    "    list for indexing to EXCLUDE outliers, remember to index with the inverse (~outliers) of the outlier\n",
    "    list returned by this function'''\n",
    "    M = adata.obs[metric]\n",
    "    outlier = ((M < np.median(M) - nmads * st.median_abs_deviation(M)) # outlier if below ...\n",
    "               | (np.median(M) + nmads * st.median_abs_deviation(M) < M)) # and outlier if above\n",
    "    return outlier\n",
    "\n",
    "# Defining whether a nucleus is an outlier based on total counts and gene counts per cell\n",
    "adata.obs[\"outlier\"] = (is_outlier(adata, \"log1p_total_counts\", 4)\n",
    "                        | is_outlier(adata, \"log1p_n_genes_by_counts\", 4)\n",
    "                        | (adata.obs[\"n_genes_by_counts\"] < 2000)) # Custom hard threshold for minimum gene number per nucleus due to Smart-seq3xpress performance\n",
    "\n",
    "# Running doublet detection and creating a doublet score for every cell\n",
    "sc.pp.scrublet(adata)\n",
    "\n",
    "print(f\"Identified {len(adata.obs[adata.obs[\"outlier\"]])} out of {len(adata.obs)} nuclei as outliers.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d786a418",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Quality Control metrics\n",
    "fig, axes = plt.subplots(3, 2, figsize=(fig_width*2, fig_width*3), layout= \"constrained\")\n",
    "\n",
    "# Plotting number of distinct genes per nucleus\n",
    "sc.pl.violin(adata, [\"n_genes_by_counts\"], jitter= 0.4, ax= axes[1, 0], show= False, palette = my_cmap.colors)\n",
    "axes[1, 0].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[1, 0].set_ylabel(\"Number of Distinct Genes\")\n",
    "axes[1, 0].axhline(adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].min(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 0].axhline(adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].max(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 0].text(-0.4, adata[~adata.obs[\"outlier\"]].obs[\"n_genes_by_counts\"].max()+170, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Plotting number of total genes per nucleus\n",
    "sc.pl.violin(adata, [\"total_counts\"], jitter= 0.4, ax= axes[1, 1], show= False, palette = my_cmap.colors)\n",
    "axes[1, 1].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[1, 1].set_ylabel(\"Total UMI Counts\")\n",
    "axes[1, 1].axhline(adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].min(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 1].axhline(adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].max(),\n",
    "                   color= my_cmap.colors[1],\n",
    "                   linestyle= \"--\",\n",
    "                   zorder= 3)\n",
    "axes[1, 1].text(-0.4, adata[~adata.obs[\"outlier\"]].obs[\"total_counts\"].max()+2000, \"outlier cutoff\",\n",
    "                color= my_cmap.colors[1],\n",
    "                fontsize= 6*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Plotting mitochondrial percentage\n",
    "sc.pl.violin(adata, [\"pct_counts_hb\"], jitter= 0.4, ax= axes[2, 0], show= False, palette = my_cmap.colors)\n",
    "axes[2, 0].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[2, 0].set_ylabel(\"Percentage of Hemoglobin Genes\")\n",
    "axes[2, 0].set_ylim(0, 0.5)\n",
    "axes[2, 0].text(-0.4, 0.485, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Plotting ribosomal percentage\n",
    "sc.pl.violin(adata, [\"pct_counts_ribo\"], jitter= 0.4, ax= axes[2, 1], show= False, palette = my_cmap.colors)\n",
    "axes[2, 1].set_xticklabels([\"Single Nuclei\"])\n",
    "axes[2, 1].set_ylabel(\"Percentages of Ribosomal Genes\")\n",
    "axes[2, 1].set_ylim(0, 0.5)\n",
    "axes[2, 1].text(-0.4, 0.485, \"y-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Plottinge highest expressed genes across whole data set\n",
    "sc.pl.highest_expr_genes(adata, n_top=20, ax= axes[0, 1], show= False, palette= my_cmap.colors)\n",
    "axes[0, 1].set_title(\"Highest Expressed Genes Total\")\n",
    "axes[0, 1].set_xlim(0, 10)\n",
    "axes[0, 1].set_xlabel(r\"% of Total Counts\")\n",
    "axes[0, 1].text(9.6, 19, \"x-axis cut for visibility\",\n",
    "                color= \"black\",\n",
    "                fontsize= 4*scale,\n",
    "                rotation= 90,\n",
    "                fontstyle= \"italic\")\n",
    "\n",
    "# Plotting histogram of count distribution\n",
    "sns.histplot(adata.obs[\"total_counts\"], bins = 100, kde= False, ax= axes[0, 0], color= my_cmap.colors[0])\n",
    "axes[0, 0].set_title(\"Distribution of Total Counts\")\n",
    "axes[0, 0].set_xlabel(\"Total UMI Counts\")\n",
    "axes[0, 0].set_ylabel(\"Number of Nuclei\")\n",
    "\n",
    "fig.tight_layout(h_pad= 1.0*scale)\n",
    "fig.savefig(figures_folder + r\"QC_Metrics_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42fe61e7",
   "metadata": {},
   "source": [
    "### Normalization\n",
    "\n",
    "We normalize counts using analytical pearson residuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d46bab85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving a copy of the raw counts before normalising\n",
    "adata.layers[\"raw_counts\"] = adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57dd1eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalizing the counts\n",
    "# ... by pearson residulals ...\n",
    "adata.layers[\"pearson_normalized_counts\"] = sc.experimental.pp.normalize_pearson_residuals(adata, copy= True).X\n",
    "\n",
    "# ... and by median (with log1p transformation)\n",
    "adata.layers[\"log1p_normalized_counts\"] = sc.pp.normalize_total(adata, copy= True).X\n",
    "sc.pp.log1p(adata, layer= \"log1p_normalized_counts\")\n",
    "\n",
    "# Plotting distributions after normalization\n",
    "fig, axes = plt.subplots(1, 3, figsize=(fig_width*2.4, fig_width*0.8), layout= \"constrained\")\n",
    "\n",
    "# the original layer remains raw\n",
    "sns.histplot(adata.X.sum(1), bins = 100, kde= False, ax= axes[0], palette= my_cmap.colors, legend= False)\n",
    "axes[0].set_title(\"Non-normalized\")\n",
    "\n",
    "sns.histplot(adata.layers[\"pearson_normalized_counts\"].sum(1), bins = 100, kde= False, ax= axes[1], color= my_cmap.colors[0])\n",
    "axes[1].set_title(\"Analytical Pearson Residuals\")\n",
    "\n",
    "\n",
    "sns.histplot(adata.layers[\"log1p_normalized_counts\"].sum(1), bins = 100, kde= False, ax= axes[2], palette= my_cmap.colors, legend= False)\n",
    "axes[2].set_title(\"Median + Log1p\")\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(figures_folder + r\"Normalization_ReferenceAtlas.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2303aa23",
   "metadata": {},
   "source": [
    "### Feature Selection\n",
    "\n",
    "We select the top 2000 most highly variable genes (HVGs) based on pearson normalized and median normalized data and save the annotation so separate fields.\n",
    "\n",
    "**DO NOT RUN THE FOLLOWING CELL TWICE WITHOUT RELOADING THE DATA OR RESTARTING THE KERNEL**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25973734",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Feature selection based on pearson residuals (uses the raw counts by default) ...\n",
    "sc.experimental.pp.highly_variable_genes(adata, n_top_genes= 2000, batch_key= \"age\")\n",
    "\n",
    "# ... and based on log1p normalized counts\n",
    "if \"log1p_highly_variable\" in adata.obs_names.values:\n",
    "    pass\n",
    "else:\n",
    "    # This very confusing code block simply makes it so that the\n",
    "    # highly variable gene metrics calculated from the log1p shifted data are saved to their own\n",
    "    # separate fields\n",
    "    adata.var = pd.concat((adata.var,\n",
    "                        sc.pp.highly_variable_genes(adata, n_top_genes= 2000,\n",
    "                                                    layer= \"log1p_normalized_counts\",\n",
    "                                                    inplace= False).rename(\n",
    "                                                        columns={\"highly_variable\": \"log1p_highly_variable\",\n",
    "                                                                    \"means\": \"log1p_means\",\n",
    "                                                                    \"dispersions\": \"log1p_dispersions\",\n",
    "                                                                    \"dispersions_norm\": \"log1p_dispersions_norm\"}).loc[:, [\"log1p_highly_variable\", \"log1p_means\", \"log1p_dispersions\", \"log1p_dispersions_norm\"]]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4bac0b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting highly variable gene fraction\n",
    "fig, ax = plt.subplots(1,2, figsize=(fig_width*1.6, fig_width*0.8), layout= \"constrained\")\n",
    "\n",
    "sns.scatterplot(adata.var, x= \"means\", y= \"residual_variances\",\n",
    "                hue= \"highly_variable\", \n",
    "                palette= my_cmap.colors,\n",
    "                ax= ax[0], \n",
    "                s= 5,\n",
    "                edgecolor= \"black\",\n",
    "                linewidth= 0.1,\n",
    "                zorder= 2)\n",
    "ax[0].set_title(\"Highly Variable Genes Pearson\")\n",
    "ax[0].legend()\n",
    "ax[0].set_xscale(\"log\")\n",
    "ax[0].set_yscale(\"log\")\n",
    "\n",
    "sns.scatterplot(adata.var, x= \"log1p_means\", y= \"log1p_dispersions\",\n",
    "                hue= \"log1p_highly_variable\",\n",
    "                palette= my_cmap.colors,\n",
    "                ax= ax[1],\n",
    "                s= 5,\n",
    "                edgecolor= \"black\",\n",
    "                linewidth= 0.1,\n",
    "                zorder= 2)\n",
    "ax[1].set_title(\"Highly Variable Genes Log1p\")\n",
    "ax[1].set_xscale(\"log\")\n",
    "ax[1].legend()\n",
    "\n",
    "fig.savefig(figures_folder + r\"HVGs_ReferenceAtlas.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1937e8e",
   "metadata": {},
   "source": [
    "### Dimensionality Reduction\n",
    "\n",
    "We compute principal components on the pearson normalized count matrix. A homogeneous plate ID distribution speaks for minimal batch effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abc44fea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making sure raw counts are used\n",
    "adata.X = adata.layers[\"raw_counts\"].copy()\n",
    "\n",
    "# Computing PCA with pearson residuals on the raw count matrix\n",
    "sc.experimental.pp.normalize_pearson_residuals_pca(adata)\n",
    "\n",
    "# We inspect the PCA results\n",
    "fig, axes = plt.subplots(1, 3, figsize=(fig_width*3, fig_width*0.8), layout= \"constrained\")\n",
    "\n",
    "sc.pl.pca_variance_ratio(adata, show= False)\n",
    "axes[0].set_title(\"Variance Ratios\")\n",
    "\n",
    "sc.pl.pca_scatter(adata, color= \"plate\",\n",
    "                  palette= my_cmap.colors,\n",
    "                  s= 10,\n",
    "                  edgecolor= \"black\",\n",
    "                  linewidth= 0.1,\n",
    "                  ax= axes[1],\n",
    "                  show= False)\n",
    "axes[1].set_title(\"PCA Results (by Plate ID)\")\n",
    "\n",
    "sc.pl.pca_scatter(adata, color= \"total_counts\",\n",
    "                  s= 10,\n",
    "                  edgecolor= \"black\",\n",
    "                  linewidth= 0.1,\n",
    "                  ax= axes[2],\n",
    "                  show= False)\n",
    "axes[2].set_title(\"PCA Results by Total Counts\")\n",
    "\n",
    "fig.savefig(figures_folder + r\"PCA_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5052ce8c",
   "metadata": {},
   "source": [
    "### UMAP Embedding\n",
    "\n",
    "We batch correct the data by calculating a batch-balances k-nearest neighbor (BBKNN) graph and calculate a UMAP projection.\n",
    "The batch covariate is the plateID, so we correct for differences between each plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "351cee41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making sure normalized counts are used\n",
    "adata.X = adata.layers[\"pearson_normalized_counts\"].copy()\n",
    "\n",
    "# BBKNN embedding\n",
    "bbknn.bbknn(adata, batch_key= \"plate\", neighbors_within_batch= 3)\n",
    "\n",
    "# UMAP Embedding\n",
    "# The min_dist and spread parameters are just for visuals\n",
    "sc.tl.umap(adata, min_dist= 1, spread= 1.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b21144",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting before filtering\n",
    "fig, axs = plt.subplots(2, 3, figsize= (fig_width*3.2, fig_width*1.6), layout= \"compressed\")\n",
    "\n",
    "metrics = [\"total_counts\", \"sample\", \"predicted_doublet\", \"n_genes_by_counts\", \"age\", \"plate\"]\n",
    "dot_size = 6\n",
    "vmax = \"p99\"\n",
    "for i, metric in enumerate(metrics):\n",
    "    sc.pl.umap(adata,\n",
    "            color=[metric],\n",
    "            palette= my_cmap.colors,\n",
    "            size= dot_size,\n",
    "            vmax= vmax,\n",
    "            edgecolor= \"black\",\n",
    "            linewidth= 0.1,\n",
    "            ax= axs[int(i/(len(metrics)/2)), int(i%(len(metrics)/2))],\n",
    "            show= False)\n",
    "axs[0, 0].text(-16,20, f\"n = {adata.n_obs}\", size= 8)\n",
    "\n",
    "fig.savefig(figures_folder + r\"UMAP_Prefilter_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7901831",
   "metadata": {},
   "source": [
    "### Filtering\n",
    "\n",
    "We filter out cells and genes that we determined to be outliers as defined above.<br>\n",
    "Exclusion criteria:\n",
    "- total genes counts outside of 4 MADs in positive or negative direction\n",
    "- less than 2000 genes expressed per nucleus (due to Smart-seq3xpress performance)\n",
    "- single genes expressed in less than 3 cells in data set\n",
    "- all mitochondrial genes (already filtered out in this file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "664fbb41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# And this is also what we have to do with out custom \"outlier\" annotations\n",
    "print(f\"Number of nuclei before outliers: {adata.n_obs}\")\n",
    "adata = adata[(~adata.obs[\"outlier\"])]\n",
    "print(f\"Number of nuclei after outliers: {adata.n_obs}\\n\")\n",
    "\n",
    "print(f\"Number of nuclei before count exclusion: {len(adata.obs)}\")\n",
    "sc.pp.filter_genes(adata, min_cells= 3, inplace= True)\n",
    "print(f\"Number of cells after count exclusion: {len(adata.obs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08258bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting after filtering\n",
    "fig, axs = plt.subplots(2, 3, figsize= (fig_width*3.2, fig_width*1.6), layout= \"compressed\")\n",
    "\n",
    "metrics = [\"total_counts\", \"sample\", \"predicted_doublet\", \"n_genes_by_counts\", \"age\", \"plate\"]\n",
    "dot_size = 6\n",
    "vmax = \"p99\"\n",
    "for i, metric in enumerate(metrics):\n",
    "    sc.pl.umap(adata,\n",
    "            color=[metric],\n",
    "            palette= my_cmap.colors,\n",
    "            size= dot_size,\n",
    "            vmax= vmax,\n",
    "            edgecolor= \"black\",\n",
    "            linewidth= 0.1,\n",
    "            ax= axs[int(i/(len(metrics)/2)), int(i%(len(metrics)/2))],\n",
    "            show= False)\n",
    "axs[0, 0].text(-16,20, f\"n = {adata.n_obs}\", size= 8)\n",
    "\n",
    "fig.savefig(figures_folder + r\"UMAP_Postfilter_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a86057f6",
   "metadata": {},
   "source": [
    "### Clustering\n",
    "\n",
    "We cluster using the Leiden algorithm at an initial resolution of 2.6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eeef10c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers[\"pearson_normalized_counts\"].copy() # The leiden() function requires normalized counts\n",
    "\n",
    "# Clustering ...\n",
    "sc.tl.leiden(adata,\n",
    "             flavor= \"igraph\",\n",
    "             n_iterations= -1, # iterate until the best clusters have been found\n",
    "             resolution= 2.6, # higher number means finer clusters\n",
    "             key_added= \"leiden\")\n",
    "\n",
    "# Plotting the clusters\n",
    "fig, ax = plt.subplots(1,1, figsize= (fig_width*0.9, fig_width*0.9))\n",
    "\n",
    "sc.pl.umap(adata, color= [\"leiden\"],\n",
    "           palette= my_cmap.colors,\n",
    "           size= 6,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1,\n",
    "           legend_loc= \"on data\",\n",
    "           legend_fontsize = 3,\n",
    "           legend_fontoutline = 0.4,\n",
    "           title= \"Leiden Clustering (resolution 2.5)\",\n",
    "           ax= ax,\n",
    "           show= False)\n",
    "\n",
    "fig.savefig(figures_folder + r\"Leiden_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c55bdeb",
   "metadata": {},
   "source": [
    "### Annotation\n",
    "\n",
    "We define a list of marker genes for defining cluster identity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a69abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Marker genes\n",
    "markers = {\"Hair Cells\": [\"Myo6\", \"Myo7a\", \"Cib2\", \"Ocm\", \"Pou4f3\"],\n",
    "           \"Inner Hair Cells\": [\"Slc17a8\", \"Otof\"],\n",
    "           \"Outer Hair Cells\": [\"Slc26a5\", \"Ikzf2\", \"Kcnq4\", \"Strc\"],\n",
    "           \"Deiter Cells\": [\"Crhr1\", \"Gdf6\", \"Plxnb2\", \"Fabp3\", \"Fgfr3\", \"Ceacam16\"],\n",
    "           \"Inner Pillar Cells\": [\"Fgfr3\", \"Lgr5\", \"Bmp6\"], # Bmp6 negative marker\n",
    "           \"Outer Pillar Cells\": [\"Fgfr3\", \"Cd44\"],\n",
    "           \"Inner Phalangeal Cells\": [\"Slc1a3\", \"Cpa6\", \"Far1\", \"Otog\", \"Gpc2\"],\n",
    "           \"Claudius Cells\": [\"Aqp4\", \"Cobl\", \"Rspo2\", \"Car12\", \"Otoa\"], # Car12 negative marker\n",
    "           \"Hensen Cells\": [\"Zbtb20\", \"Lockd\", \"Dpysl4\", \"Cobl\", \"Cd44\"], # Cd44 negative marker\n",
    "           \"Boettcher Cells\": [\"Car12\", \"Rspo2\"],\n",
    "           \"Inner Sulcus Cells\": [\"Aqp4\", \"Cd44 \", \"Bmp4\"],\n",
    "           \"Interdental Cells\": [\"Otoa\", \"Ceacam16\"],\n",
    "           \"Macrophages\": [\"Fcgr3\", \"Csf1r\", \"Lgmn\", \"C1qb\", \"Mrc1\", \"Ptprc\", \"C1qb\"],\n",
    "           \"Endothelial Cells\": [\"Flt1\", \"Tie1\", \"Ly6c\", \"Esam\", \"Cdh5\", \"Pecam\"],\n",
    "           \"Pericytes\": [\"Rgs5\", \"Pdgfrb\", \"Cspg4\"],\n",
    "           \"Intermediate Cells\": [\"Kcnj10\", \"Dct\", \"Col5a3\"],\n",
    "           \"Marginal Cells\": [\"Kcne1\", \"Bsnd\"],\n",
    "           \"Basal Cells\": [\"Cldn11\", \"Atp6v0a4\"],\n",
    "           \"Spindle Cells\": [\"Slc26a4\", \"Anxa1\", \"Dpp10\"], #Slc26a4 strong in both Spindle and Root Cells\n",
    "           \"Root Cells\": [\"Slc26a4\", \"Lgr5\", \"Epyc\"],\n",
    "           \"Reissner Cells\": [\"Lmx1a\", \"Frem2\", \"Slc26a7\"],\n",
    "           \"Tympanic Border Cells\": [\"Emilin2\", \"Rarres1\", \"Emcn\", \"Otor\"],\n",
    "           \"Osteoblast\": [\"Runx2\", \"Satb2\", \"Dio2\", \"Dlx5\", \"Bglap\"],\n",
    "           \"Osteocyte\": [\"Phex\", \"Sfrp4\", \"Fxyd5\"],\n",
    "           \"Chondrocytes\": [\"Prg4\", \"Fn1\", \"Clu\"],\n",
    "           \"Fibrocytes\": [\"Coch\", \"Otos\", \"Car3\", \"Col1a2\", \"Slco1c1\"],\n",
    "           \"Fibroblast\": [\"Dcn\", \"Lum\"],\n",
    "           \"Fibrocyte I\": [\"Car2\", \"Col9a2\", \"Slc4a11\"],\n",
    "           \"Fibrocyte II\": [\"Atp1a2\"],\n",
    "           \"Fibrocyte III\": [\"Cald1\", \"Aqp1\"],\n",
    "           \"Fibrocyte IV\": [\"Ranbp3l\", \"Ptgds\"],\n",
    "           \"Neurons\": [\"Snap25\", \"Isl5\", \"Rbfox3\", \"Prox1\"],\n",
    "           \"Spiral Ganglion Neuron Type I\": [\"Snap25\", \"Tuj1\", \"Prox1\", \"Fos\", \"Nrn1\"],\n",
    "           \"Spiral Ganglion Neuron Type Ia\": [\"Calb2\", \"B3gat1\", \"Rxrg\"],\n",
    "           \"Spiral Ganglion Neuron Type Ib\": [\"Ntng1\", \"Runx1\", \"Ttn\", \"Calb1\"],\n",
    "           \"Spiral Ganglion Neuron Type Ic\": [\"Lypd1\", \"Grm8\", \"Pou4f1\"],\n",
    "           \"Spiral Ganglion Neuron Type II\": [\"Gata3\", \"Pcdh20\", \"Prph\", \"Nf200\", \"Plk5\"],\n",
    "           \"Oligodendrocytes\": [\"Mog\", \"Plp1\", \"Tubb4a\", \"Mobp\"],\n",
    "           \"Myelinating Schwann Cells\": [\"Ncmap\", \"Mpz\", \"Drp2\", \"Prx\", \"Kcna1\", \"Kcna2\", \"Pmp22\"], #Kcna1 and Kcna2 strong for both Schwann Cell Types\n",
    "           \"Non-myelinating Schwann Cells\": [\"Scn7a\", \"Cdh19\", \"Lgi4\", \"Kcna1\", \"Kcna2\", \"Kcna6 \"],\n",
    "           \"Satellite Glial Cells\": [\"Aqp4\", \"Gja1\", \"Ptprz1\", \"Bcan\", \"Fbln2 \", \"Igfbp6\"]}\n",
    "\n",
    "# Saving marker genes for later\n",
    "with open(r\"Resources\\cochlea_markers.json\", \"w\") as file:\n",
    "    json.dump(markers, file)\n",
    "\n",
    "# Loading previously saved markers\n",
    "with open(r\"Resources\\cochlea_markers.json\", \"r\") as file:\n",
    "    markers = json.load(file)\n",
    "\n",
    "# Checking which marker genes are present in the dataset\n",
    "markers_in_data = {}\n",
    "for ct, markers in markers.items():\n",
    "    if np.intersect1d(markers, adata.var_names).size > 0: # We check whether the marker name is included in the gene names of the matrix\n",
    "        markers_in_data[ct] = list(np.intersect1d(markers, adata.var_names))\n",
    "\n",
    "# printing the final list of available marker genes\n",
    "print(\"Marker genes present in data:\\n\")\n",
    "for item in markers_in_data.items():\n",
    "    print(f\"{item[0]}: {item[1]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3d17e10",
   "metadata": {},
   "source": [
    "### Checking Marker Expression\n",
    "\n",
    "We visualize the expression of the marker genes in each cluster and infer which clusters could belong to which cell type.\n",
    "\n",
    "Initally we just observe if the markers locate to clusters unambiguously, if not we sub-cluster the clusters that are too coarse in the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89c4a7c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dotplot featuring all marker genes and all the clusters to manually determine cell identities in clusters\n",
    "# THIS IS WHERE THE ACTUAL CLUSTER ANNOTATION CALLING HAPPENS\n",
    "sc.pl.dotplot(adata,\n",
    "              markers_in_data,\n",
    "              groupby=\"leiden_0\",\n",
    "              standard_scale=\"var\",\n",
    "              cmap= \"Purples\", # I like purple\n",
    "            #   save= \"Marker_Genes_ReferenceAtlas.png\", # save the dotplot, comment out if desired\n",
    "              )\n",
    "\n",
    "# Plotting leiden clusters for cross referencing\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"leiden_0\"],\n",
    "           legend_loc=\"on data\",\n",
    "           legend_fontsize= 2,\n",
    "           legend_fontoutline= 0.8,\n",
    "           size= 6,\n",
    "           palette= my_cmap.colors,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79effe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# After initial inspection of clustering, we defining clusters that look like they wanna be clustered further\n",
    "# This is iterative sub-clustering\n",
    "need_subclustering = {\"7\": 1, # \"cluster\": [subclustering resolution]\n",
    "                     \"17\": 1,\n",
    "                     \"19\": 1,\n",
    "                     \"1\": 0.5,\n",
    "                     \"20\": 1,\n",
    "                     \"26\": 1}\n",
    "\n",
    "# Successive sub-clustering of cluters that are too coarse to tell cell types apart\n",
    "i = 1\n",
    "for cluster, res in need_subclustering.items():\n",
    "    sc.tl.leiden(adata, flavor= \"igraph\", n_iterations= -1, key_added= f\"leiden_{i}\",\n",
    "                 resolution= res,\n",
    "                 restrict_to= (f\"leiden_{i-1}\", [cluster]))\n",
    "    i += 1\n",
    "\n",
    "# Dotplot featuring all marker genes and all the clusters to manually determine cell identities in clusters\n",
    "# THIS IS WHERE THE ACTUAL CLUSTER ANNOTATION CALLING HAPPENS\n",
    "sc.pl.dotplot(adata,\n",
    "              markers_in_data,\n",
    "              groupby=f\"leiden_{len(need_subclustering)}\",\n",
    "              standard_scale=\"var\",\n",
    "              cmap= \"Purples\", # I like purple\n",
    "            #   save= \"Marker_Genes_ReferenceAtlas.png\", # save the dotplot, comment out if desired\n",
    "              )\n",
    "\n",
    "# Plotting leiden clusters for cross referencing\n",
    "sc.pl.umap(adata,\n",
    "           color=[f\"leiden_{len(need_subclustering)}\"],\n",
    "           legend_loc=\"on data\",\n",
    "           legend_fontsize= 2,\n",
    "           legend_fontoutline= 0.8,\n",
    "           size= 6,\n",
    "           palette= my_cmap.colors,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cb37d15",
   "metadata": {},
   "source": [
    "### Checking Cluster Expressions\n",
    "\n",
    "We visualise marker gene expressions on the UMAP projection to visually verify localizations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a8e893d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining a subgroup of cell types of which to check expression levels across the whole data set\n",
    "# Change cell types as needed!\n",
    "sgns = [\n",
    "    \"Spiral Ganglion Neuron Type Ia\",\n",
    "    \"Spiral Ganglion Neuron Type Ib\",\n",
    "    \"Spiral Ganglion Neuron Type Ic\",\n",
    "    \"Spiral Ganglion Neuron Type II\"]\n",
    "\n",
    "# Now we plot marker gene expression for each marker of the above defined cell types\n",
    "for ct in sgns:\n",
    "    print(f\"{ct}:\")  # print cell subtype name\n",
    "    sc.pl.umap(adata,\n",
    "               color= [*markers_in_data[ct], f\"leiden_{len(need_subclustering)}\"],\n",
    "               vmax=\"p99\",\n",
    "               sort_order=False,  # do not plot highest expression on top, to not get a biased view of the mean expression among cell,\n",
    "               frameon=False,\n",
    "               size= 6,\n",
    "               palette= my_cmap.colors,\n",
    "               edgecolor= \"black\",\n",
    "               linewidth= 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99cc068",
   "metadata": {},
   "source": [
    "### Comitting Annotations\n",
    "\n",
    "We now define a manual cell type for each numeric cluster from before and map the annotated cells to a new field in the .obs annotations (every nucleus gets a cell type label)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e4f1db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Committing Annotations\n",
    "annotations = {\"0\": \"Myelinating Schwann Cells 1\",\n",
    "                           \"1,0\": \"Fibrocytes IV\",\n",
    "                           \"1,1\": \"Fibrocytes IV\",\n",
    "                           \"1,2\": \"Fibrocytes II\",\n",
    "                           \"2\": \"Fibrocytes IV\",\n",
    "                           \"3\": \"Myelinating Schwann Cells 1\",\n",
    "                           \"4\": \"Fibroblasts\",\n",
    "                           \"5\": \"Osteocytes\",\n",
    "                           \"6\": \"Chondrocytes\",\n",
    "                           \"7,1\": \"Spiral Ganglion Neuron Type Ia\",\n",
    "                           \"7,3\": \"Spiral Ganglion Neuron Type Ia\",\n",
    "                           \"7,0\": \"Spiral Ganglion Neuron Type Ib\",\n",
    "                           \"7,4\": \"Spiral Ganglion Neuron Type Ib\",\n",
    "                           \"7,2\": \"Spiral Ganglion Neuron Type Ic\",\n",
    "                           \"7,5\": \"Spiral Ganglion Neuron Type II\",\n",
    "                           \"8\": \"Fibrocytes III\",\n",
    "                           \"9\": \"Endothelial Cells\",\n",
    "                           \"10\": \"Oligodendrocytes\",\n",
    "                           \"11\": \"Satellite Glial Cells 1\",\n",
    "                           \"12\": \"Macrophages\",\n",
    "                           \"13\": \"Non-myelinating Schwann Cells\",\n",
    "                           \"14\": \"Pericytes\",\n",
    "                           \"15\": \"Fibrocytes I\",\n",
    "                           \"16\": \"Basal Cells\",\n",
    "                           \"17,0\": \"Inner Hair Cells\",\n",
    "                           \"17,1\": \"Outer Hair Cells\",\n",
    "                           \"17,2\": \"Outer Hair Cells\",\n",
    "                           \"18\": \"Basal Cells\",\n",
    "                           \"19,0\": \"Root Cells\",\n",
    "                           \"19,1\": \"Hensen Cells\",\n",
    "                           \"19,2\": \"Hensen Cells\",\n",
    "                           \"20,0\": \"Claudius Cells\",\n",
    "                           \"20,1\": \"Claudius Cells\",\n",
    "                           \"20,2\": \"Interdental Cells\",\n",
    "                           \"20,3\": \"Claudius Cells\",\n",
    "                           \"20,4\": \"Interdental Cells\",\n",
    "                           \"20,5\": \"Claudius Cells\",\n",
    "                           \"21\": \"Myelinating Schwann Cells 2\",\n",
    "                           \"22\": \"Reissner Cells\",\n",
    "                           \"23\": \"Intermediate Cells\",\n",
    "                           \"24\": \"Tympanic Border Cells\",\n",
    "                           \"25\": \"Marginal Cells\",\n",
    "                           \"26,0\": \"Pillar Cells\",\n",
    "                           \"26,1\": \"Pillar Cells\",\n",
    "                           \"26,2\": \"Deiter Cells\",\n",
    "                           \"27\": \"Spindle Cells\",\n",
    "                           \"28\": \"Inner Phalangeal Cells\",\n",
    "                           \"29\": \"Marginal Cells\",\n",
    "                           \"30\": \"Intermediate Cells\",\n",
    "                           \"31\": \"Satellite Glial Cells 1\",\n",
    "                           \"32\": \"Satellite Glial Cells 2\"}\n",
    "\n",
    "# Refactoring the cluster identity and saving it to a new annotation as cell types\n",
    "adata.obs[\"manual_celltypes\"] = adata.obs[\"leiden_6\"].map(annotations)\n",
    "\n",
    "# Saving the annotated dataset\n",
    "adata.write_h5ad(data_folder + r\"CBA_3mo_WholeCochlea_Annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d996054a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting our pretty little clusters\n",
    "fig, ax = plt.subplots(1,1, figsize=(fig_width*2.5, fig_width), layout= \"compressed\")\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"manual_celltypes\"],\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.15,\n",
    "           size= 10,\n",
    "           palette= my_cmap.colors,\n",
    "           legend_fontsize= 6*scale,\n",
    "           ax = ax,\n",
    "           show= False)\n",
    "ax.set_title(\"Whole Cochlea Reference Atlas\")\n",
    "\n",
    "fig.savefig(figures_folder + r\"Annotations_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1c72b72",
   "metadata": {},
   "source": [
    "## Gene Expression Analysis\n",
    "\n",
    "Here, as an example, we filter out only the spiral ganglion neurons and plot the expression of one selected marker for each cluster on top of the UMAP projection.<br>\n",
    "\n",
    "For this we reload the annotated file as to not have to rerun all the above cells when reanalysing the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f73bdd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading previously annotated data\n",
    "adata = sc.read_h5ad(data_folder + r\"CBA_3mo_WholeCochlea_Annotated.h5ad\")\n",
    "\n",
    "# Filtering the whole dataframe to only contain SGNs (Type II is included)\n",
    "adata_SGNonly = adata[adata.obs[\"manual_celltypes\"].str.contains(\"Type I\")].copy()\n",
    "markers_to_plot = [\"Calb2\", # Type Ia\n",
    "                   \"Ttn\",   # Type Ib\n",
    "                   \"Lypd1\", # Type Ic\n",
    "                   \"Plk5\"]  # Type II\n",
    "\n",
    "# Plotting expressions of SGN marker genes in only the SGN cluster\n",
    "fig, axes = plt.subplots(5,2, figsize=(fig_width*1.9, fig_width*2.85),\n",
    "                         layout= \"compressed\",\n",
    "                         gridspec_kw= {\"wspace\": 0.1})\n",
    "\n",
    "dot_size = 50\n",
    "line_width = 0.5\n",
    "vmax = \"p99\"\n",
    "sc.pl.umap(adata_SGNonly,\n",
    "           color=[\"manual_celltypes\"],\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= line_width,\n",
    "           size= dot_size,\n",
    "           palette= my_cmap.colors[30:34],\n",
    "           ax= axes[0,1],\n",
    "           show= False,\n",
    "           vmax= vmax,\n",
    "           title= \"\")\n",
    "\n",
    "for i, mark in enumerate(markers_to_plot):\n",
    "    sc.pl.umap(adata_SGNonly,\n",
    "               color= mark,\n",
    "               size= dot_size,\n",
    "               vmax= vmax,\n",
    "               legend_loc= \"none\",\n",
    "               edgecolor= \"black\",\n",
    "               linewidth= line_width,\n",
    "               ax= axes[i+1, 0],\n",
    "               title= \"\",\n",
    "               show= False)\n",
    "    axes[i+1, 0].set_title(mark, fontsize= 8*scale)\n",
    "    \n",
    "    sc.pl.violin(adata_SGNonly,\n",
    "                 mark,\n",
    "                 groupby= \"manual_celltypes\",\n",
    "                 jitter= 0.4,\n",
    "                 size= 2,\n",
    "                 palette= my_cmap.colors[30:34],\n",
    "                 ax= axes[i+1, 1],\n",
    "                 rotation= 90,\n",
    "                 ylabel= \"\",\n",
    "                 zorder= 2,\n",
    "                 show= False)\n",
    "    axes[i+1, 1].set_xticklabels([\"\"])\n",
    "    axes[i+1, 1].set_title(\"\")\n",
    "    axes[i+1, 1].set_ylabel(\"Normalized Expression\")\n",
    "else:\n",
    "    axes[i+1, 1].set_xticklabels([\"Type II\", \"Type Ia\", \"Type Ib\", \"Type Ic\"])\n",
    "\n",
    "fig.savefig(figures_folder + r\"SGN_MarkerExpression_ReferenceAtlas.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "330999f6",
   "metadata": {},
   "source": [
    "### Unsupervised Gene Expression Analysis\n",
    "\n",
    "Here we rank the top 5 differentially expressed genes for each of the manually annotated clusters. This is a way of potentially finding new markers and verifyfing the expression of existing markers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6106886",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata,\n",
    "                        groupby=\"manual_celltypes\", # The gene groups key\n",
    "                        method=\"wilcoxon\",\n",
    "                        n_genes= 5)\n",
    "\n",
    "sc.pl.rank_genes_groups_dotplot(adata,\n",
    "                                groupby=\"manual_celltypes\",\n",
    "                                standard_scale=\"var\",\n",
    "                                cmap= \"Purples\",\n",
    "                                # save= \"GeneRanks_ReferenceAtlas.png\", # comment in if desired\n",
    "                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42e89861",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we can visualize these unsupervised found genes across the whole dataset\n",
    "cell_type = [\"Spiral Ganglion Neuron Type Ia\"]\n",
    "new_markers = sc.get.rank_genes_groups_df(adata, group= cell_type)\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color= new_markers[\"names\"].to_list(),\n",
    "           sort_order=False,\n",
    "           size= 6,\n",
    "           edgecolor= \"black\",\n",
    "           linewidth= 0.1)\n",
    "\n",
    "new_markers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5894cd91",
   "metadata": {},
   "source": [
    "### END OF SCRIPT\n",
    "This notebook was created by *Johann Korn* on 01-10-2025."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
